// clay-management-toolbar/lib/ClayManagementToolbar.soy.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.templates = exports.ClayManagementToolbar = undefined;

var _metalComponent = require('metal-component');

var _metalComponent2 = _interopRequireDefault(_metalComponent);

var _metalSoy = require('metal-soy');

var _metalSoy2 = _interopRequireDefault(_metalSoy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /* jshint ignore:start */


var templates;
goog.loadModule(function (exports) {
  var soy = goog.require('soy');
  var soydata = goog.require('soydata');
  // This file was automatically generated from ClayManagementToolbar.soy.
  // Please don't edit this file by hand.

  /**
   * @fileoverview Templates in namespace ClayManagementToolbar.
   * @hassoydeltemplate {ClayManagementToolbar.SearchForm.idom}
   * @hassoydelcall {ClayManagementToolbar.SearchForm.idom}
   * @public
   */

  goog.module('ClayManagementToolbar.incrementaldom');

  goog.require('goog.soy.data.SanitizedContent');
  goog.require('goog.string');
  var incrementalDom = goog.require('incrementaldom');
  goog.require('soy');
  goog.require('soy.asserts');
  var soyIdom = goog.require('soy.idom');

  var $templateAlias5 = _metalSoy2.default.getTemplate('ClayActionsDropdown.incrementaldom', 'render');

  var $templateAlias4 = _metalSoy2.default.getTemplate('ClayButton.incrementaldom', 'render');

  var $templateAlias2 = _metalSoy2.default.getTemplate('ClayCheckbox.incrementaldom', 'render');

  var $templateAlias6 = _metalSoy2.default.getTemplate('ClayCreationMenuDropdown.incrementaldom', 'render');

  var $templateAlias9 = _metalSoy2.default.getTemplate('ClayDropdown.incrementaldom', 'render');

  var $templateAlias8 = _metalSoy2.default.getTemplate('ClayDropdownBase.incrementaldom', 'render');

  var $templateAlias7 = _metalSoy2.default.getTemplate('ClayIcon.incrementaldom', 'render');

  var $templateAlias3 = _metalSoy2.default.getTemplate('ClayLink.incrementaldom', 'render');

  var $templateAlias1 = _metalSoy2.default.getTemplate('ClayResultsBar.incrementaldom', 'render');

  /**
   * @param {{
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleActionItemClicked: (*|null|undefined),
   *  _handleClearResultsClick: (*|null|undefined),
   *  _handleClearSelectionButtonClick: (*|null|undefined),
   *  _handleCloseMobileSearchClick: (*|null|undefined),
   *  _handleCreationButtonClicked: (*|null|undefined),
   *  _handleCreationMenuItemClicked: (*|null|undefined),
   *  _handleCreationMenuMoreButtonClicked: (*|null|undefined),
   *  _handleFilterDoneButtonClick: (*|null|undefined),
   *  _handleFilterItemClicked: (*|null|undefined),
   *  _handleFilterLabelCloseClicked: (*|null|undefined),
   *  _handleInfoButtonClicked: (*|null|undefined),
   *  _handleOpenMobileSearchClick: (*|null|undefined),
   *  _handleQuickActionClicked: (*|null|undefined),
   *  _handleSearchSearchClick: (*|null|undefined),
   *  _handleSelectAllButtonClick: (*|null|undefined),
   *  _handleSelectPageCheckboxChanged: (*|null|undefined),
   *  _handleSortingButtonClicked: (*|null|undefined),
   *  _handleViewTypeClicked: (*|null|undefined),
   *  _showSearchMobile: (boolean|null|undefined),
   *  actionItems: (!Array<{disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), quickAction: boolean, separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined),
   *  active: (boolean|null|undefined),
   *  checkboxStatus: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  clearResultsURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  clearSelectionURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  creationMenu: (null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)}),
   *  customData: (?),
   *  disabled: (boolean|null|undefined),
   *  elementClasses: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  filterItems: (!Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), items: !Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined),
   *  filterLabels: (!Array<{closeable: boolean, label: (!goog.soy.data.SanitizedContent|function()|string)}>|null|undefined),
   *  id: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchActionURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormMethod: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInputPlaceholder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchValue: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  selectable: (boolean|null|undefined),
   *  selectAllURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  selectedItems: (null|number|undefined),
   *  showAdvancedSearch: (boolean|null|undefined),
   *  showCreationMenu: (boolean|null|undefined),
   *  showFiltersDoneButton: (boolean|null|undefined),
   *  showInfoButton: (boolean|null|undefined),
   *  showResultsBar: (boolean|null|undefined),
   *  showSearch: (boolean|null|undefined),
   *  showSelectAllButton: (boolean|null|undefined),
   *  sortingOrder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  sortingURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  supportsBulkActions: (boolean|null|undefined),
   *  totalItems: (null|number|undefined),
   *  viewTypes: (!Array<{active: boolean, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string)}>|null|undefined)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function $render(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    var $$temp;
    /** @type {!goog.soy.data.SanitizedContent|string} */
    var spritemap = soy.asserts.assertType(goog.isString(opt_data.spritemap) || opt_data.spritemap instanceof goog.soy.data.SanitizedContent, 'spritemap', opt_data.spritemap, '!goog.soy.data.SanitizedContent|string');
    /** @type {*|null|undefined} */
    var _handleActionItemClicked = opt_data._handleActionItemClicked;
    /** @type {*|null|undefined} */
    var _handleClearResultsClick = opt_data._handleClearResultsClick;
    /** @type {*|null|undefined} */
    var _handleClearSelectionButtonClick = opt_data._handleClearSelectionButtonClick;
    /** @type {*|null|undefined} */
    var _handleCloseMobileSearchClick = opt_data._handleCloseMobileSearchClick;
    /** @type {*|null|undefined} */
    var _handleCreationButtonClicked = opt_data._handleCreationButtonClicked;
    /** @type {*|null|undefined} */
    var _handleCreationMenuItemClicked = opt_data._handleCreationMenuItemClicked;
    /** @type {*|null|undefined} */
    var _handleCreationMenuMoreButtonClicked = opt_data._handleCreationMenuMoreButtonClicked;
    /** @type {*|null|undefined} */
    var _handleFilterDoneButtonClick = opt_data._handleFilterDoneButtonClick;
    /** @type {*|null|undefined} */
    var _handleFilterItemClicked = opt_data._handleFilterItemClicked;
    /** @type {*|null|undefined} */
    var _handleFilterLabelCloseClicked = opt_data._handleFilterLabelCloseClicked;
    /** @type {*|null|undefined} */
    var _handleInfoButtonClicked = opt_data._handleInfoButtonClicked;
    /** @type {*|null|undefined} */
    var _handleOpenMobileSearchClick = opt_data._handleOpenMobileSearchClick;
    /** @type {*|null|undefined} */
    var _handleQuickActionClicked = opt_data._handleQuickActionClicked;
    /** @type {*|null|undefined} */
    var _handleSearchSearchClick = opt_data._handleSearchSearchClick;
    /** @type {*|null|undefined} */
    var _handleSelectAllButtonClick = opt_data._handleSelectAllButtonClick;
    /** @type {*|null|undefined} */
    var _handleSelectPageCheckboxChanged = opt_data._handleSelectPageCheckboxChanged;
    /** @type {*|null|undefined} */
    var _handleSortingButtonClicked = opt_data._handleSortingButtonClicked;
    /** @type {*|null|undefined} */
    var _handleViewTypeClicked = opt_data._handleViewTypeClicked;
    /** @type {boolean|null|undefined} */
    var _showSearchMobile = soy.asserts.assertType(opt_data._showSearchMobile == null || goog.isBoolean(opt_data._showSearchMobile) || opt_data._showSearchMobile === 1 || opt_data._showSearchMobile === 0, '_showSearchMobile', opt_data._showSearchMobile, 'boolean|null|undefined');
    /** @type {!Array<{disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), quickAction: boolean, separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined} */
    var actionItems = soy.asserts.assertType(opt_data.actionItems == null || goog.isArray(opt_data.actionItems), 'actionItems', opt_data.actionItems, '!Array<{disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), quickAction: boolean, separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined');
    /** @type {boolean|null|undefined} */
    var active = soy.asserts.assertType(opt_data.active == null || goog.isBoolean(opt_data.active) || opt_data.active === 1 || opt_data.active === 0, 'active', opt_data.active, 'boolean|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var checkboxStatus = soy.asserts.assertType(opt_data.checkboxStatus == null || goog.isString(opt_data.checkboxStatus) || opt_data.checkboxStatus instanceof goog.soy.data.SanitizedContent, 'checkboxStatus', opt_data.checkboxStatus, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var clearResultsURL = soy.asserts.assertType(opt_data.clearResultsURL == null || goog.isString(opt_data.clearResultsURL) || opt_data.clearResultsURL instanceof goog.soy.data.SanitizedContent, 'clearResultsURL', opt_data.clearResultsURL, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var clearSelectionURL = soy.asserts.assertType(opt_data.clearSelectionURL == null || goog.isString(opt_data.clearSelectionURL) || opt_data.clearSelectionURL instanceof goog.soy.data.SanitizedContent, 'clearSelectionURL', opt_data.clearSelectionURL, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var contentRenderer = soy.asserts.assertType(opt_data.contentRenderer == null || goog.isString(opt_data.contentRenderer) || opt_data.contentRenderer instanceof goog.soy.data.SanitizedContent, 'contentRenderer', opt_data.contentRenderer, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)}} */
    var creationMenu = soy.asserts.assertType(opt_data.creationMenu == null || goog.isObject(opt_data.creationMenu), 'creationMenu', opt_data.creationMenu, 'null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)}');
    /** @type {?} */
    var customData = opt_data.customData;
    /** @type {boolean|null|undefined} */
    var disabled = soy.asserts.assertType(opt_data.disabled == null || goog.isBoolean(opt_data.disabled) || opt_data.disabled === 1 || opt_data.disabled === 0, 'disabled', opt_data.disabled, 'boolean|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var elementClasses = soy.asserts.assertType(opt_data.elementClasses == null || goog.isString(opt_data.elementClasses) || opt_data.elementClasses instanceof goog.soy.data.SanitizedContent, 'elementClasses', opt_data.elementClasses, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), items: !Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined} */
    var filterItems = soy.asserts.assertType(opt_data.filterItems == null || goog.isArray(opt_data.filterItems), 'filterItems', opt_data.filterItems, '!Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), items: !Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined');
    /** @type {!Array<{closeable: boolean, label: (!goog.soy.data.SanitizedContent|function()|string)}>|null|undefined} */
    var filterLabels = soy.asserts.assertType(opt_data.filterLabels == null || goog.isArray(opt_data.filterLabels), 'filterLabels', opt_data.filterLabels, '!Array<{closeable: boolean, label: (!goog.soy.data.SanitizedContent|function()|string)}>|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var id = soy.asserts.assertType(opt_data.id == null || goog.isString(opt_data.id) || opt_data.id instanceof goog.soy.data.SanitizedContent, 'id', opt_data.id, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchActionURL = soy.asserts.assertType(opt_data.searchActionURL == null || goog.isString(opt_data.searchActionURL) || opt_data.searchActionURL instanceof goog.soy.data.SanitizedContent, 'searchActionURL', opt_data.searchActionURL, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchFormMethod = soy.asserts.assertType(opt_data.searchFormMethod == null || goog.isString(opt_data.searchFormMethod) || opt_data.searchFormMethod instanceof goog.soy.data.SanitizedContent, 'searchFormMethod', opt_data.searchFormMethod, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchFormName = soy.asserts.assertType(opt_data.searchFormName == null || goog.isString(opt_data.searchFormName) || opt_data.searchFormName instanceof goog.soy.data.SanitizedContent, 'searchFormName', opt_data.searchFormName, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchInputName = soy.asserts.assertType(opt_data.searchInputName == null || goog.isString(opt_data.searchInputName) || opt_data.searchInputName instanceof goog.soy.data.SanitizedContent, 'searchInputName', opt_data.searchInputName, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchInputPlaceholder = soy.asserts.assertType(opt_data.searchInputPlaceholder == null || goog.isString(opt_data.searchInputPlaceholder) || opt_data.searchInputPlaceholder instanceof goog.soy.data.SanitizedContent, 'searchInputPlaceholder', opt_data.searchInputPlaceholder, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchValue = soy.asserts.assertType(opt_data.searchValue == null || goog.isString(opt_data.searchValue) || opt_data.searchValue instanceof goog.soy.data.SanitizedContent, 'searchValue', opt_data.searchValue, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {boolean|null|undefined} */
    var selectable = soy.asserts.assertType(opt_data.selectable == null || goog.isBoolean(opt_data.selectable) || opt_data.selectable === 1 || opt_data.selectable === 0, 'selectable', opt_data.selectable, 'boolean|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var selectAllURL = soy.asserts.assertType(opt_data.selectAllURL == null || goog.isString(opt_data.selectAllURL) || opt_data.selectAllURL instanceof goog.soy.data.SanitizedContent, 'selectAllURL', opt_data.selectAllURL, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {null|number|undefined} */
    var selectedItems = soy.asserts.assertType(opt_data.selectedItems == null || goog.isNumber(opt_data.selectedItems), 'selectedItems', opt_data.selectedItems, 'null|number|undefined');
    /** @type {boolean|null|undefined} */
    var showAdvancedSearch = soy.asserts.assertType(opt_data.showAdvancedSearch == null || goog.isBoolean(opt_data.showAdvancedSearch) || opt_data.showAdvancedSearch === 1 || opt_data.showAdvancedSearch === 0, 'showAdvancedSearch', opt_data.showAdvancedSearch, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showCreationMenu = soy.asserts.assertType(opt_data.showCreationMenu == null || goog.isBoolean(opt_data.showCreationMenu) || opt_data.showCreationMenu === 1 || opt_data.showCreationMenu === 0, 'showCreationMenu', opt_data.showCreationMenu, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showFiltersDoneButton = soy.asserts.assertType(opt_data.showFiltersDoneButton == null || goog.isBoolean(opt_data.showFiltersDoneButton) || opt_data.showFiltersDoneButton === 1 || opt_data.showFiltersDoneButton === 0, 'showFiltersDoneButton', opt_data.showFiltersDoneButton, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showInfoButton = soy.asserts.assertType(opt_data.showInfoButton == null || goog.isBoolean(opt_data.showInfoButton) || opt_data.showInfoButton === 1 || opt_data.showInfoButton === 0, 'showInfoButton', opt_data.showInfoButton, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showResultsBar = soy.asserts.assertType(opt_data.showResultsBar == null || goog.isBoolean(opt_data.showResultsBar) || opt_data.showResultsBar === 1 || opt_data.showResultsBar === 0, 'showResultsBar', opt_data.showResultsBar, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showSearch = soy.asserts.assertType(opt_data.showSearch == null || goog.isBoolean(opt_data.showSearch) || opt_data.showSearch === 1 || opt_data.showSearch === 0, 'showSearch', opt_data.showSearch, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showSelectAllButton = soy.asserts.assertType(opt_data.showSelectAllButton == null || goog.isBoolean(opt_data.showSelectAllButton) || opt_data.showSelectAllButton === 1 || opt_data.showSelectAllButton === 0, 'showSelectAllButton', opt_data.showSelectAllButton, 'boolean|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var sortingOrder = soy.asserts.assertType(opt_data.sortingOrder == null || goog.isString(opt_data.sortingOrder) || opt_data.sortingOrder instanceof goog.soy.data.SanitizedContent, 'sortingOrder', opt_data.sortingOrder, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var sortingURL = soy.asserts.assertType(opt_data.sortingURL == null || goog.isString(opt_data.sortingURL) || opt_data.sortingURL instanceof goog.soy.data.SanitizedContent, 'sortingURL', opt_data.sortingURL, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {boolean|null|undefined} */
    var supportsBulkActions = soy.asserts.assertType(opt_data.supportsBulkActions == null || goog.isBoolean(opt_data.supportsBulkActions) || opt_data.supportsBulkActions === 1 || opt_data.supportsBulkActions === 0, 'supportsBulkActions', opt_data.supportsBulkActions, 'boolean|null|undefined');
    /** @type {null|number|undefined} */
    var totalItems = soy.asserts.assertType(opt_data.totalItems == null || goog.isNumber(opt_data.totalItems), 'totalItems', opt_data.totalItems, 'null|number|undefined');
    /** @type {!Array<{active: boolean, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string)}>|null|undefined} */
    var viewTypes = soy.asserts.assertType(opt_data.viewTypes == null || goog.isArray(opt_data.viewTypes), 'viewTypes', opt_data.viewTypes, '!Array<{active: boolean, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string)}>|null|undefined');
    var attributes__soy107 = function attributes__soy107() {
      if (elementClasses) {
        incrementalDom.attr('class', elementClasses);
      }
      if (id) {
        incrementalDom.attr('id', id);
      }
    };
    var isActive__soy122 = active != null ? active : checkboxStatus && (checkboxStatus == 'checked' || checkboxStatus == 'indeterminate');
    incrementalDom.elementOpenStart('div');
    attributes__soy107();
    incrementalDom.elementOpenEnd();
    if (isActive__soy122) {
      $active({ _handleActionItemClicked: _handleActionItemClicked, _handleClearSelectionButtonClick: _handleClearSelectionButtonClick, _handleInfoButtonClicked: _handleInfoButtonClicked, _handleQuickActionClicked: _handleQuickActionClicked, _handleSelectAllButtonClick: _handleSelectAllButtonClick, _handleSelectPageCheckboxChanged: _handleSelectPageCheckboxChanged, actionItems: ($$temp = actionItems) == null ? [] : $$temp, checkboxStatus: checkboxStatus, clearSelectionURL: clearSelectionURL, selectAllURL: selectAllURL, selectedItems: selectedItems, showInfoButton: showInfoButton, showSelectAllButton: showSelectAllButton, spritemap: spritemap, supportsBulkActions: supportsBulkActions, totalItems: totalItems }, null, opt_ijData);
    } else {
      $default({ _handleClearResultsClick: _handleClearResultsClick, _handleCloseMobileSearchClick: _handleCloseMobileSearchClick, _handleCreationButtonClicked: _handleCreationButtonClicked, _handleCreationMenuItemClicked: _handleCreationMenuItemClicked, _handleCreationMenuMoreButtonClicked: _handleCreationMenuMoreButtonClicked, _handleFilterDoneButtonClick: _handleFilterDoneButtonClick, _handleFilterItemClicked: _handleFilterItemClicked, _handleInfoButtonClicked: _handleInfoButtonClicked, _handleOpenMobileSearchClick: _handleOpenMobileSearchClick, _handleSearchSearchClick: _handleSearchSearchClick, _handleSelectPageCheckboxChanged: _handleSelectPageCheckboxChanged, _handleSortingButtonClicked: _handleSortingButtonClicked, _handleViewTypeClicked: _handleViewTypeClicked, _showSearchMobile: _showSearchMobile, contentRenderer: ($$temp = contentRenderer) == null ? '' : $$temp, creationMenu: creationMenu, customData: customData, disabled: disabled, filterItems: filterItems, searchActionURL: searchActionURL, searchFormMethod: searchFormMethod, searchFormName: searchFormName, searchInputName: searchInputName, searchInputPlaceholder: searchInputPlaceholder, searchValue: searchValue, selectable: selectable, showAdvancedSearch: showAdvancedSearch, showCreationMenu: ($$temp = showCreationMenu) == null ? true : $$temp, showFiltersDoneButton: ($$temp = showFiltersDoneButton) == null ? true : $$temp, showInfoButton: showInfoButton, showSearch: showSearch != false, sortingOrder: sortingOrder, sortingURL: sortingURL, spritemap: spritemap, viewTypes: viewTypes }, null, opt_ijData);
    }
    if (showResultsBar) {
      $templateAlias1({ _handleClearResultsClick: _handleClearResultsClick, clearResultsURL: clearResultsURL, events: { filterLabelCloseClicked: _handleFilterLabelCloseClicked }, filterLabels: ($$temp = filterLabels) == null ? [] : $$temp, ref: 'resultsBar', searchValue: searchValue, spritemap: spritemap, totalItems: ($$temp = totalItems) == null ? 0 : $$temp }, null, opt_ijData);
    }
    incrementalDom.elementClose('div');
  }
  exports.render = $render;
  /**
   * @typedef {{
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleActionItemClicked: (*|null|undefined),
   *  _handleClearResultsClick: (*|null|undefined),
   *  _handleClearSelectionButtonClick: (*|null|undefined),
   *  _handleCloseMobileSearchClick: (*|null|undefined),
   *  _handleCreationButtonClicked: (*|null|undefined),
   *  _handleCreationMenuItemClicked: (*|null|undefined),
   *  _handleCreationMenuMoreButtonClicked: (*|null|undefined),
   *  _handleFilterDoneButtonClick: (*|null|undefined),
   *  _handleFilterItemClicked: (*|null|undefined),
   *  _handleFilterLabelCloseClicked: (*|null|undefined),
   *  _handleInfoButtonClicked: (*|null|undefined),
   *  _handleOpenMobileSearchClick: (*|null|undefined),
   *  _handleQuickActionClicked: (*|null|undefined),
   *  _handleSearchSearchClick: (*|null|undefined),
   *  _handleSelectAllButtonClick: (*|null|undefined),
   *  _handleSelectPageCheckboxChanged: (*|null|undefined),
   *  _handleSortingButtonClicked: (*|null|undefined),
   *  _handleViewTypeClicked: (*|null|undefined),
   *  _showSearchMobile: (boolean|null|undefined),
   *  actionItems: (!Array<{disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), quickAction: boolean, separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined),
   *  active: (boolean|null|undefined),
   *  checkboxStatus: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  clearResultsURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  clearSelectionURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  creationMenu: (null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)}),
   *  customData: (?),
   *  disabled: (boolean|null|undefined),
   *  elementClasses: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  filterItems: (!Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), items: !Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined),
   *  filterLabels: (!Array<{closeable: boolean, label: (!goog.soy.data.SanitizedContent|function()|string)}>|null|undefined),
   *  id: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchActionURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormMethod: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInputPlaceholder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchValue: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  selectable: (boolean|null|undefined),
   *  selectAllURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  selectedItems: (null|number|undefined),
   *  showAdvancedSearch: (boolean|null|undefined),
   *  showCreationMenu: (boolean|null|undefined),
   *  showFiltersDoneButton: (boolean|null|undefined),
   *  showInfoButton: (boolean|null|undefined),
   *  showResultsBar: (boolean|null|undefined),
   *  showSearch: (boolean|null|undefined),
   *  showSelectAllButton: (boolean|null|undefined),
   *  sortingOrder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  sortingURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  supportsBulkActions: (boolean|null|undefined),
   *  totalItems: (null|number|undefined),
   *  viewTypes: (!Array<{active: boolean, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string)}>|null|undefined)
   * }}
   */
  $render.Params;
  if (goog.DEBUG) {
    $render.soyTemplateName = 'ClayManagementToolbar.render';
  }

  /**
   * @param {{
   *  actionItems: !Array<?>,
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleActionItemClicked: (*|null|undefined),
   *  _handleClearSelectionButtonClick: (*|null|undefined),
   *  _handleInfoButtonClicked: (*|null|undefined),
   *  _handleQuickActionClicked: (*|null|undefined),
   *  _handleSelectAllButtonClick: (*|null|undefined),
   *  _handleSelectPageCheckboxChanged: (*|null|undefined),
   *  checkboxStatus: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  clearSelectionURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  selectAllURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  selectedItems: (null|number|undefined),
   *  showInfoButton: (boolean|null|undefined),
   *  showSelectAllButton: (boolean|null|undefined),
   *  supportsBulkActions: (boolean|null|undefined),
   *  totalItems: (null|number|undefined)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function $active(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    /** @type {!Array<?>} */
    var actionItems = soy.asserts.assertType(goog.isArray(opt_data.actionItems), 'actionItems', opt_data.actionItems, '!Array<?>');
    /** @type {!goog.soy.data.SanitizedContent|string} */
    var spritemap = soy.asserts.assertType(goog.isString(opt_data.spritemap) || opt_data.spritemap instanceof goog.soy.data.SanitizedContent, 'spritemap', opt_data.spritemap, '!goog.soy.data.SanitizedContent|string');
    /** @type {*|null|undefined} */
    var _handleActionItemClicked = opt_data._handleActionItemClicked;
    /** @type {*|null|undefined} */
    var _handleClearSelectionButtonClick = opt_data._handleClearSelectionButtonClick;
    /** @type {*|null|undefined} */
    var _handleInfoButtonClicked = opt_data._handleInfoButtonClicked;
    /** @type {*|null|undefined} */
    var _handleQuickActionClicked = opt_data._handleQuickActionClicked;
    /** @type {*|null|undefined} */
    var _handleSelectAllButtonClick = opt_data._handleSelectAllButtonClick;
    /** @type {*|null|undefined} */
    var _handleSelectPageCheckboxChanged = opt_data._handleSelectPageCheckboxChanged;
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var checkboxStatus = soy.asserts.assertType(opt_data.checkboxStatus == null || goog.isString(opt_data.checkboxStatus) || opt_data.checkboxStatus instanceof goog.soy.data.SanitizedContent, 'checkboxStatus', opt_data.checkboxStatus, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var clearSelectionURL = soy.asserts.assertType(opt_data.clearSelectionURL == null || goog.isString(opt_data.clearSelectionURL) || opt_data.clearSelectionURL instanceof goog.soy.data.SanitizedContent, 'clearSelectionURL', opt_data.clearSelectionURL, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var selectAllURL = soy.asserts.assertType(opt_data.selectAllURL == null || goog.isString(opt_data.selectAllURL) || opt_data.selectAllURL instanceof goog.soy.data.SanitizedContent, 'selectAllURL', opt_data.selectAllURL, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {null|number|undefined} */
    var selectedItems = soy.asserts.assertType(opt_data.selectedItems == null || goog.isNumber(opt_data.selectedItems), 'selectedItems', opt_data.selectedItems, 'null|number|undefined');
    /** @type {boolean|null|undefined} */
    var showInfoButton = soy.asserts.assertType(opt_data.showInfoButton == null || goog.isBoolean(opt_data.showInfoButton) || opt_data.showInfoButton === 1 || opt_data.showInfoButton === 0, 'showInfoButton', opt_data.showInfoButton, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showSelectAllButton = soy.asserts.assertType(opt_data.showSelectAllButton == null || goog.isBoolean(opt_data.showSelectAllButton) || opt_data.showSelectAllButton === 1 || opt_data.showSelectAllButton === 0, 'showSelectAllButton', opt_data.showSelectAllButton, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var supportsBulkActions = soy.asserts.assertType(opt_data.supportsBulkActions == null || goog.isBoolean(opt_data.supportsBulkActions) || opt_data.supportsBulkActions === 1 || opt_data.supportsBulkActions === 0, 'supportsBulkActions', opt_data.supportsBulkActions, 'boolean|null|undefined');
    /** @type {null|number|undefined} */
    var totalItems = soy.asserts.assertType(opt_data.totalItems == null || goog.isNumber(opt_data.totalItems), 'totalItems', opt_data.totalItems, 'null|number|undefined');
    var msg_4409__soy4410 = '';
    /** @desc Select items checkbox label */
    var MSG_EXTERNAL_2532519427604455359 = goog.getMsg('select-items');
    msg_4409__soy4410 += MSG_EXTERNAL_2532519427604455359;
    incrementalDom.elementOpenStart('nav');
    incrementalDom.attr('class', 'management-bar management-bar-primary navbar navbar-expand-md navbar-nowrap');
    incrementalDom.elementOpenEnd();
    incrementalDom.elementOpenStart('div');
    incrementalDom.attr('class', 'container-fluid container-fluid-max-xl');
    incrementalDom.elementOpenEnd();
    incrementalDom.elementOpenStart('ul');
    incrementalDom.attr('class', 'navbar-nav navbar-nav-expand');
    incrementalDom.elementOpenEnd();
    incrementalDom.elementOpenStart('li');
    incrementalDom.attr('class', 'nav-item');
    incrementalDom.elementOpenEnd();
    $templateAlias2({ checked: checkboxStatus != 'unchecked', events: { change: _handleSelectPageCheckboxChanged }, indeterminate: checkboxStatus == 'indeterminate', label: '' + msg_4409__soy4410, ref: 'selectPageCheckbox', showLabel: false }, null, opt_ijData);
    incrementalDom.elementClose('li');
    incrementalDom.elementOpenStart('li');
    incrementalDom.attr('class', 'nav-item');
    incrementalDom.elementOpenEnd();
    incrementalDom.elementOpenStart('span');
    incrementalDom.attr('class', 'navbar-text');
    incrementalDom.elementOpenEnd();
    if (selectedItems == totalItems) {
      /** @desc Says that all possible items are selected */
      var MSG_EXTERNAL_3292565618977302987 = goog.getMsg('all-selected');
      incrementalDom.text(goog.string.unescapeEntities(MSG_EXTERNAL_3292565618977302987));
    } else {
      /** @desc Says how many items of the total are selected */
      var MSG_EXTERNAL_65349286544398424 = goog.getMsg('{$selectedItems}-of-{$totalItems}', { 'selectedItems': '\x010\x01', 'totalItems': '\x011\x01' });
      var lastIndex_244 = 0,
          partRe_244 = /\x01\d+\x01/g,
          match_244;
      do {
        match_244 = partRe_244.exec(MSG_EXTERNAL_65349286544398424) || undefined;
        incrementalDom.text(goog.string.unescapeEntities(MSG_EXTERNAL_65349286544398424.substring(lastIndex_244, match_244 && match_244.index)));
        lastIndex_244 = partRe_244.lastIndex;
        switch (match_244 && match_244[0]) {
          case '\x011\x01':
            soyIdom.print(totalItems);
            break;
          case '\x010\x01':
            soyIdom.print(selectedItems);
            break;
        }
      } while (match_244);
      incrementalDom.text(' ');
      incrementalDom.elementOpenStart('span');
      incrementalDom.attr('class', 'navbar-breakpoint-down-d-none');
      incrementalDom.elementOpenEnd();
      /** @desc Items selected */
      var MSG_EXTERNAL_3024397635961501335 = goog.getMsg('selected');
      incrementalDom.text(goog.string.unescapeEntities(MSG_EXTERNAL_3024397635961501335));
      incrementalDom.elementClose('span');
    }
    incrementalDom.elementClose('span');
    incrementalDom.elementClose('li');
    if (supportsBulkActions) {
      incrementalDom.elementOpenStart('li');
      incrementalDom.attr('class', 'nav-item nav-item-shrink');
      incrementalDom.elementOpenEnd();
      if (clearSelectionURL) {
        var param269 = function param269() {
          incrementalDom.elementOpenStart('span');
          incrementalDom.attr('class', 'text-truncate-inline');
          incrementalDom.elementOpenEnd();
          incrementalDom.elementOpenStart('span');
          incrementalDom.attr('class', 'text-truncate');
          incrementalDom.elementOpenEnd();
          /** @desc Clear */
          var MSG_EXTERNAL_517843663917688276 = goog.getMsg('clear');
          incrementalDom.text(goog.string.unescapeEntities(MSG_EXTERNAL_517843663917688276));
          incrementalDom.elementClose('span');
          incrementalDom.elementClose('span');
        };
        $templateAlias3({ elementClasses: 'nav-link', href: clearSelectionURL, label: param269 }, null, opt_ijData);
      } else {
        var param282 = function param282() {
          incrementalDom.elementOpenStart('span');
          incrementalDom.attr('class', 'text-truncate-inline');
          incrementalDom.elementOpenEnd();
          incrementalDom.elementOpenStart('span');
          incrementalDom.attr('class', 'text-truncate');
          incrementalDom.elementOpenEnd();
          /** @desc Clear */
          var MSG_EXTERNAL_517843663917688276$$1 = goog.getMsg('clear');
          incrementalDom.text(goog.string.unescapeEntities(MSG_EXTERNAL_517843663917688276$$1));
          incrementalDom.elementClose('span');
          incrementalDom.elementClose('span');
        };
        $templateAlias4({ elementClasses: 'nav-link', events: { click: _handleClearSelectionButtonClick }, label: param282, ref: 'clearSelectionButton', style: 'unstyled', type: 'button' }, null, opt_ijData);
      }
      incrementalDom.elementClose('li');
      if (showSelectAllButton) {
        incrementalDom.elementOpenStart('li');
        incrementalDom.attr('class', 'nav-item nav-item-shrink');
        incrementalDom.elementOpenEnd();
        if (selectAllURL) {
          var param303 = function param303() {
            incrementalDom.elementOpenStart('span');
            incrementalDom.attr('class', 'text-truncate-inline');
            incrementalDom.elementOpenEnd();
            incrementalDom.elementOpenStart('span');
            incrementalDom.attr('class', 'text-truncate');
            incrementalDom.elementOpenEnd();
            /** @desc  */
            var MSG_EXTERNAL_7670214129603586399 = goog.getMsg('select-all');
            incrementalDom.text(goog.string.unescapeEntities(MSG_EXTERNAL_7670214129603586399));
            incrementalDom.elementClose('span');
            incrementalDom.elementClose('span');
          };
          $templateAlias3({ elementClasses: 'nav-link', href: selectAllURL, label: param303 }, null, opt_ijData);
        } else {
          var param316 = function param316() {
            incrementalDom.elementOpenStart('span');
            incrementalDom.attr('class', 'text-truncate-inline');
            incrementalDom.elementOpenEnd();
            incrementalDom.elementOpenStart('span');
            incrementalDom.attr('class', 'text-truncate');
            incrementalDom.elementOpenEnd();
            /** @desc  */
            var MSG_EXTERNAL_7670214129603586399$$1 = goog.getMsg('select-all');
            incrementalDom.text(goog.string.unescapeEntities(MSG_EXTERNAL_7670214129603586399$$1));
            incrementalDom.elementClose('span');
            incrementalDom.elementClose('span');
          };
          $templateAlias4({ elementClasses: 'nav-link', events: { click: _handleSelectAllButtonClick }, label: param316, ref: 'selectAllButton', style: 'unstyled', type: 'button' }, null, opt_ijData);
        }
        incrementalDom.elementClose('li');
      }
    }
    incrementalDom.elementClose('ul');
    var hasActionItems__soy330 = actionItems.length > 0;
    if (hasActionItems__soy330 || showInfoButton) {
      incrementalDom.elementOpenStart('ul');
      incrementalDom.attr('class', 'navbar-nav');
      incrementalDom.elementOpenEnd();
      if (showInfoButton) {
        var msg_4412__soy4413 = '';
        /** @desc  */
        var MSG_EXTERNAL_4279081882680795350 = goog.getMsg('info');
        msg_4412__soy4413 += MSG_EXTERNAL_4279081882680795350;
        incrementalDom.elementOpenStart('li');
        incrementalDom.attr('class', 'nav-item');
        incrementalDom.elementOpenEnd();
        $templateAlias4({ elementClasses: 'nav-link nav-link-monospaced', events: { click: _handleInfoButtonClicked }, icon: 'info-circle-open', ref: 'infoButton', spritemap: spritemap, style: 'unstyled', title: '' + msg_4412__soy4413 }, null, opt_ijData);
        incrementalDom.elementClose('li');
      }
      if (hasActionItems__soy330) {
        var item398List = actionItems;
        var item398ListLen = item398List.length;
        for (var item398Index = 0; item398Index < item398ListLen; item398Index++) {
          var item398Data = item398List[item398Index];
          if (item398Data.quickAction && item398Data.icon && spritemap) {
            var attributes__soy361 = function attributes__soy361() {
              incrementalDom.attr('class', 'nav-item navbar-breakpoint-down-d-none');
              if (!item398Data.disabled) {
                incrementalDom.attr('data-onclick', _handleQuickActionClicked);
              }
              incrementalDom.attr('data-quick-action-index', item398Index);
            };
            incrementalDom.elementOpenStart('li');
            attributes__soy361();
            incrementalDom.elementOpenEnd();
            if (item398Data.href != null && item398Data.href != '' && (!(item398Data.disabled != null) || item398Data.disabled == false)) {
              $templateAlias3({ elementClasses: 'nav-link nav-link-monospaced', href: item398Data.href, icon: item398Data.icon, ref: 'quickAction' + item398Index, spritemap: spritemap, title: item398Data.label }, null, opt_ijData);
            } else {
              $templateAlias4({ disabled: item398Data.disabled, elementClasses: 'nav-link nav-link-monospaced', icon: item398Data.icon, ref: 'quickAction' + item398Index, spritemap: spritemap, style: 'unstyled', title: item398Data.label }, null, opt_ijData);
            }
            incrementalDom.elementClose('li');
          }
        }
        incrementalDom.elementOpenStart('li');
        incrementalDom.attr('class', 'nav-item');
        incrementalDom.elementOpenEnd();
        $templateAlias5({ events: { itemClicked: _handleActionItemClicked }, items: actionItems, spritemap: spritemap, triggerClasses: 'nav-link nav-link-monospaced' }, null, opt_ijData);
        incrementalDom.elementClose('li');
      }
      incrementalDom.elementClose('ul');
    }
    incrementalDom.elementClose('div');
    incrementalDom.elementClose('nav');
  }
  exports.active = $active;
  /**
   * @typedef {{
   *  actionItems: !Array<?>,
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleActionItemClicked: (*|null|undefined),
   *  _handleClearSelectionButtonClick: (*|null|undefined),
   *  _handleInfoButtonClicked: (*|null|undefined),
   *  _handleQuickActionClicked: (*|null|undefined),
   *  _handleSelectAllButtonClick: (*|null|undefined),
   *  _handleSelectPageCheckboxChanged: (*|null|undefined),
   *  checkboxStatus: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  clearSelectionURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  selectAllURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  selectedItems: (null|number|undefined),
   *  showInfoButton: (boolean|null|undefined),
   *  showSelectAllButton: (boolean|null|undefined),
   *  supportsBulkActions: (boolean|null|undefined),
   *  totalItems: (null|number|undefined)
   * }}
   */
  $active.Params;
  if (goog.DEBUG) {
    $active.soyTemplateName = 'ClayManagementToolbar.active';
  }

  /**
   * @param {{
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleCreationButtonClicked: (*|null|undefined),
   *  _handleCreationMenuItemClicked: (*|null|undefined),
   *  _handleCreationMenuMoreButtonClicked: (*|null|undefined),
   *  creationMenu: (null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)})
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function $creationMenu(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    /** @type {!goog.soy.data.SanitizedContent|string} */
    var spritemap = soy.asserts.assertType(goog.isString(opt_data.spritemap) || opt_data.spritemap instanceof goog.soy.data.SanitizedContent, 'spritemap', opt_data.spritemap, '!goog.soy.data.SanitizedContent|string');
    /** @type {*|null|undefined} */
    var _handleCreationButtonClicked = opt_data._handleCreationButtonClicked;
    /** @type {*|null|undefined} */
    var _handleCreationMenuItemClicked = opt_data._handleCreationMenuItemClicked;
    /** @type {*|null|undefined} */
    var _handleCreationMenuMoreButtonClicked = opt_data._handleCreationMenuMoreButtonClicked;
    /** @type {null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)}} */
    var creationMenu = soy.asserts.assertType(opt_data.creationMenu == null || goog.isObject(opt_data.creationMenu), 'creationMenu', opt_data.creationMenu, 'null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)}');
    var primaryItems__soy416 = creationMenu && creationMenu.primaryItems ? creationMenu.primaryItems : [];
    var secondaryItems__soy418 = creationMenu && creationMenu.secondaryItems ? creationMenu.secondaryItems : [];
    var hasPrimaryItems__soy420 = primaryItems__soy416.length > 0;
    var hasSecondaryItems__soy422 = secondaryItems__soy418.length > 0;
    var firstSecondaryItem__soy424 = hasSecondaryItems__soy422 ? secondaryItems__soy418[0] : null;
    var secondSecondaryItem__soy426 = hasSecondaryItems__soy422 ? secondaryItems__soy418[1] : null;
    var hasFirstSecondaryItemNestedItems__soy428 = firstSecondaryItem__soy424 && firstSecondaryItem__soy424.items && firstSecondaryItem__soy424.items.length > 0;
    var hasSecondSecondaryItemNestedItems__soy430 = secondSecondaryItem__soy426 && secondSecondaryItem__soy426.items && secondSecondaryItem__soy426.items.length > 0;
    var hasOnePrimaryItem__soy432 = hasPrimaryItems__soy420 && primaryItems__soy416.length == 1;
    var hasOneSecondaryItem__soy434 = hasSecondaryItems__soy422 && secondaryItems__soy418.length == 1 && !hasFirstSecondaryItemNestedItems__soy428 && !hasSecondSecondaryItemNestedItems__soy430 || hasFirstSecondaryItemNestedItems__soy428 && firstSecondaryItem__soy424.items.length == 1 && !hasSecondSecondaryItemNestedItems__soy430 || hasSecondSecondaryItemNestedItems__soy430 && secondSecondaryItem__soy426.items.length == 1 && !hasFirstSecondaryItemNestedItems__soy428;
    if (hasOnePrimaryItem__soy432 && !hasSecondaryItems__soy422 || hasOneSecondaryItem__soy434 && !hasPrimaryItems__soy420) {
      var item__soy439 = primaryItems__soy416 && primaryItems__soy416[0] ? primaryItems__soy416[0] : secondaryItems__soy418[0].items ? secondaryItems__soy418[0].items[0] : secondaryItems__soy418[0];
      if (item__soy439.href != null && item__soy439.href != '') {
        $templateAlias3({ ariaLabel: item__soy439.label, buttonStyle: 'primary', data: item__soy439.data, elementClasses: 'nav-btn nav-btn-monospaced', href: item__soy439.href, icon: 'plus', spritemap: spritemap, title: item__soy439.label }, null, opt_ijData);
      } else {
        $templateAlias4({ data: item__soy439.data, elementClasses: 'nav-btn nav-btn-monospaced', events: { click: _handleCreationButtonClicked }, icon: 'plus', ref: 'creationMenu', spritemap: spritemap, title: item__soy439.label }, null, opt_ijData);
      }
    } else if (hasPrimaryItems__soy420 || hasSecondaryItems__soy422) {
      $templateAlias6({ caption: creationMenu.caption, events: { itemClicked: _handleCreationMenuItemClicked, moreButtonClicked: _handleCreationMenuMoreButtonClicked }, helpText: creationMenu.helpText, itemsIconAlignment: creationMenu.itemsIconAlignment, maxPrimaryItems: creationMenu.maxPrimaryItems, maxSecondaryItems: creationMenu.maxSecondaryItems, maxTotalItems: creationMenu.maxTotalItems, primaryItems: creationMenu.primaryItems, ref: 'creationMenuDropdown', secondaryItems: creationMenu.secondaryItems, spritemap: spritemap, triggerClasses: 'nav-btn nav-btn-monospaced', viewMoreURL: creationMenu.viewMoreURL }, null, opt_ijData);
    } else {
      var msg_4415__soy4416 = '';
      /** @desc  */
      var MSG_EXTERNAL_7387635272539030076 = goog.getMsg('new');
      msg_4415__soy4416 += MSG_EXTERNAL_7387635272539030076;
      $templateAlias4({ elementClasses: 'nav-btn nav-btn-monospaced', events: { click: _handleCreationButtonClicked }, icon: 'plus', ref: 'creationMenu', spritemap: spritemap, title: '' + msg_4415__soy4416 }, null, opt_ijData);
    }
  }
  exports.creationMenu = $creationMenu;
  /**
   * @typedef {{
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleCreationButtonClicked: (*|null|undefined),
   *  _handleCreationMenuItemClicked: (*|null|undefined),
   *  _handleCreationMenuMoreButtonClicked: (*|null|undefined),
   *  creationMenu: (null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)})
   * }}
   */
  $creationMenu.Params;
  if (goog.DEBUG) {
    $creationMenu.soyTemplateName = 'ClayManagementToolbar.creationMenu';
  }

  /**
   * @param {{
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleCloseMobileSearchClick: (*|null|undefined),
   *  _handleCreationButtonClicked: (*|null|undefined),
   *  _handleCreationMenuItemClicked: (*|null|undefined),
   *  _handleCreationMenuMoreButtonClicked: (*|null|undefined),
   *  _handleFilterDoneButtonClick: (*|null|undefined),
   *  _handleFilterItemClicked: (*|null|undefined),
   *  _handleInfoButtonClicked: (*|null|undefined),
   *  _handleOpenMobileSearchClick: (*|null|undefined),
   *  _handleSearchSearchClick: (*|null|undefined),
   *  _handleSelectPageCheckboxChanged: (*|null|undefined),
   *  _handleSortingButtonClicked: (*|null|undefined),
   *  _handleViewTypeClicked: (*|null|undefined),
   *  _showSearchMobile: (boolean|null|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  creationMenu: (null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)}),
   *  customData: (?),
   *  disabled: (boolean|null|undefined),
   *  filterItems: (!Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), items: !Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined),
   *  searchActionURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormMethod: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInputPlaceholder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchValue: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  selectable: (boolean|null|undefined),
   *  showAdvancedSearch: (boolean|null|undefined),
   *  showCreationMenu: (boolean|null|undefined),
   *  showFiltersDoneButton: (boolean|null|undefined),
   *  showInfoButton: (boolean|null|undefined),
   *  showSearch: (boolean|null|undefined),
   *  sortingOrder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  sortingURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  viewTypes: (!Array<{active: boolean, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string)}>|null|undefined)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function $default(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    /** @type {!goog.soy.data.SanitizedContent|string} */
    var spritemap = soy.asserts.assertType(goog.isString(opt_data.spritemap) || opt_data.spritemap instanceof goog.soy.data.SanitizedContent, 'spritemap', opt_data.spritemap, '!goog.soy.data.SanitizedContent|string');
    /** @type {*|null|undefined} */
    var _handleCloseMobileSearchClick = opt_data._handleCloseMobileSearchClick;
    /** @type {*|null|undefined} */
    var _handleCreationButtonClicked = opt_data._handleCreationButtonClicked;
    /** @type {*|null|undefined} */
    var _handleCreationMenuItemClicked = opt_data._handleCreationMenuItemClicked;
    /** @type {*|null|undefined} */
    var _handleCreationMenuMoreButtonClicked = opt_data._handleCreationMenuMoreButtonClicked;
    /** @type {*|null|undefined} */
    var _handleFilterDoneButtonClick = opt_data._handleFilterDoneButtonClick;
    /** @type {*|null|undefined} */
    var _handleFilterItemClicked = opt_data._handleFilterItemClicked;
    /** @type {*|null|undefined} */
    var _handleInfoButtonClicked = opt_data._handleInfoButtonClicked;
    /** @type {*|null|undefined} */
    var _handleOpenMobileSearchClick = opt_data._handleOpenMobileSearchClick;
    /** @type {*|null|undefined} */
    var _handleSearchSearchClick = opt_data._handleSearchSearchClick;
    /** @type {*|null|undefined} */
    var _handleSelectPageCheckboxChanged = opt_data._handleSelectPageCheckboxChanged;
    /** @type {*|null|undefined} */
    var _handleSortingButtonClicked = opt_data._handleSortingButtonClicked;
    /** @type {*|null|undefined} */
    var _handleViewTypeClicked = opt_data._handleViewTypeClicked;
    /** @type {boolean|null|undefined} */
    var _showSearchMobile = soy.asserts.assertType(opt_data._showSearchMobile == null || goog.isBoolean(opt_data._showSearchMobile) || opt_data._showSearchMobile === 1 || opt_data._showSearchMobile === 0, '_showSearchMobile', opt_data._showSearchMobile, 'boolean|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var contentRenderer = soy.asserts.assertType(opt_data.contentRenderer == null || goog.isString(opt_data.contentRenderer) || opt_data.contentRenderer instanceof goog.soy.data.SanitizedContent, 'contentRenderer', opt_data.contentRenderer, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)}} */
    var creationMenu = soy.asserts.assertType(opt_data.creationMenu == null || goog.isObject(opt_data.creationMenu), 'creationMenu', opt_data.creationMenu, 'null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)}');
    /** @type {?} */
    var customData = opt_data.customData;
    /** @type {boolean|null|undefined} */
    var disabled = soy.asserts.assertType(opt_data.disabled == null || goog.isBoolean(opt_data.disabled) || opt_data.disabled === 1 || opt_data.disabled === 0, 'disabled', opt_data.disabled, 'boolean|null|undefined');
    /** @type {!Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), items: !Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined} */
    var filterItems = soy.asserts.assertType(opt_data.filterItems == null || goog.isArray(opt_data.filterItems), 'filterItems', opt_data.filterItems, '!Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), items: !Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchActionURL = soy.asserts.assertType(opt_data.searchActionURL == null || goog.isString(opt_data.searchActionURL) || opt_data.searchActionURL instanceof goog.soy.data.SanitizedContent, 'searchActionURL', opt_data.searchActionURL, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchFormMethod = soy.asserts.assertType(opt_data.searchFormMethod == null || goog.isString(opt_data.searchFormMethod) || opt_data.searchFormMethod instanceof goog.soy.data.SanitizedContent, 'searchFormMethod', opt_data.searchFormMethod, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchFormName = soy.asserts.assertType(opt_data.searchFormName == null || goog.isString(opt_data.searchFormName) || opt_data.searchFormName instanceof goog.soy.data.SanitizedContent, 'searchFormName', opt_data.searchFormName, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchInputName = soy.asserts.assertType(opt_data.searchInputName == null || goog.isString(opt_data.searchInputName) || opt_data.searchInputName instanceof goog.soy.data.SanitizedContent, 'searchInputName', opt_data.searchInputName, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchInputPlaceholder = soy.asserts.assertType(opt_data.searchInputPlaceholder == null || goog.isString(opt_data.searchInputPlaceholder) || opt_data.searchInputPlaceholder instanceof goog.soy.data.SanitizedContent, 'searchInputPlaceholder', opt_data.searchInputPlaceholder, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchValue = soy.asserts.assertType(opt_data.searchValue == null || goog.isString(opt_data.searchValue) || opt_data.searchValue instanceof goog.soy.data.SanitizedContent, 'searchValue', opt_data.searchValue, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {boolean|null|undefined} */
    var selectable = soy.asserts.assertType(opt_data.selectable == null || goog.isBoolean(opt_data.selectable) || opt_data.selectable === 1 || opt_data.selectable === 0, 'selectable', opt_data.selectable, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showAdvancedSearch = soy.asserts.assertType(opt_data.showAdvancedSearch == null || goog.isBoolean(opt_data.showAdvancedSearch) || opt_data.showAdvancedSearch === 1 || opt_data.showAdvancedSearch === 0, 'showAdvancedSearch', opt_data.showAdvancedSearch, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showCreationMenu = soy.asserts.assertType(opt_data.showCreationMenu == null || goog.isBoolean(opt_data.showCreationMenu) || opt_data.showCreationMenu === 1 || opt_data.showCreationMenu === 0, 'showCreationMenu', opt_data.showCreationMenu, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showFiltersDoneButton = soy.asserts.assertType(opt_data.showFiltersDoneButton == null || goog.isBoolean(opt_data.showFiltersDoneButton) || opt_data.showFiltersDoneButton === 1 || opt_data.showFiltersDoneButton === 0, 'showFiltersDoneButton', opt_data.showFiltersDoneButton, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showInfoButton = soy.asserts.assertType(opt_data.showInfoButton == null || goog.isBoolean(opt_data.showInfoButton) || opt_data.showInfoButton === 1 || opt_data.showInfoButton === 0, 'showInfoButton', opt_data.showInfoButton, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var showSearch = soy.asserts.assertType(opt_data.showSearch == null || goog.isBoolean(opt_data.showSearch) || opt_data.showSearch === 1 || opt_data.showSearch === 0, 'showSearch', opt_data.showSearch, 'boolean|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var sortingOrder = soy.asserts.assertType(opt_data.sortingOrder == null || goog.isString(opt_data.sortingOrder) || opt_data.sortingOrder instanceof goog.soy.data.SanitizedContent, 'sortingOrder', opt_data.sortingOrder, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var sortingURL = soy.asserts.assertType(opt_data.sortingURL == null || goog.isString(opt_data.sortingURL) || opt_data.sortingURL instanceof goog.soy.data.SanitizedContent, 'sortingURL', opt_data.sortingURL, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!Array<{active: boolean, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string)}>|null|undefined} */
    var viewTypes = soy.asserts.assertType(opt_data.viewTypes == null || goog.isArray(opt_data.viewTypes), 'viewTypes', opt_data.viewTypes, '!Array<{active: boolean, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string)}>|null|undefined');
    incrementalDom.elementOpenStart('nav');
    incrementalDom.attr('class', 'management-bar management-bar-light navbar navbar-expand-md');
    incrementalDom.elementOpenEnd();
    var onlySearch__soy537 = showSearch && !creationMenu && !selectable && !filterItems && !sortingOrder && !viewTypes;
    incrementalDom.elementOpenStart('div');
    incrementalDom.attr('class', 'container-fluid container-fluid-max-xl');
    incrementalDom.elementOpenEnd();
    incrementalDom.elementOpenStart('ul');
    incrementalDom.attr('class', 'navbar-nav');
    incrementalDom.elementOpenEnd();
    if (selectable) {
      var msg_4418__soy4419 = '';
      /** @desc Select items checkbox label */
      var MSG_EXTERNAL_2532519427604455359 = goog.getMsg('select-items');
      msg_4418__soy4419 += MSG_EXTERNAL_2532519427604455359;
      incrementalDom.elementOpenStart('li');
      incrementalDom.attr('class', 'nav-item');
      incrementalDom.elementOpenEnd();
      $templateAlias2({ disabled: disabled, events: { change: _handleSelectPageCheckboxChanged }, label: '' + msg_4418__soy4419, ref: 'selectPageCheckbox', showLabel: !filterItems && !sortingOrder && !showSearch }, null, opt_ijData);
      incrementalDom.elementClose('li');
    }
    if (filterItems) {
      var msg_4421__soy4422 = '';
      /** @desc Done */
      var MSG_EXTERNAL_6608890834230007808 = goog.getMsg('done');
      msg_4421__soy4422 += MSG_EXTERNAL_6608890834230007808;
      var filtersTrigger__soy559 = function filtersTrigger__soy559() {
        incrementalDom.elementOpenStart('span');
        incrementalDom.attr('class', 'navbar-breakpoint-down-d-none');
        incrementalDom.elementOpenEnd();
        /** @desc Filters and order dropdown trigger label */
        var MSG_EXTERNAL_4557218837305114639 = goog.getMsg('filter-and-order');
        incrementalDom.text(goog.string.unescapeEntities(MSG_EXTERNAL_4557218837305114639));
        $templateAlias7({ elementClasses: 'inline-item inline-item-after', spritemap: spritemap, symbol: 'caret-bottom' }, null, opt_ijData);
        incrementalDom.elementClose('span');
        incrementalDom.elementOpenStart('span');
        incrementalDom.attr('class', 'navbar-breakpoint-d-none');
        incrementalDom.elementOpenEnd();
        $templateAlias7({ spritemap: spritemap, symbol: 'filter' }, null, opt_ijData);
        incrementalDom.elementClose('span');
      };
      incrementalDom.elementOpenStart('li');
      incrementalDom.attr('class', 'dropdown nav-item');
      incrementalDom.elementOpenEnd();
      var doneString__soy576 = '';
      doneString__soy576 += msg_4421__soy4422;
      $templateAlias8({ button: showFiltersDoneButton ? { label: doneString__soy576 } : null, disabled: disabled, events: { buttonClicked: _handleFilterDoneButtonClick, itemClicked: _handleFilterItemClicked }, items: filterItems, label: filtersTrigger__soy559, ref: 'filtersDropdown', spritemap: spritemap, triggerClasses: 'nav-link' }, null, opt_ijData);
      incrementalDom.elementClose('li');
    }
    if (sortingOrder) {
      incrementalDom.elementOpenStart('li');
      incrementalDom.attr('class', 'nav-item');
      incrementalDom.elementOpenEnd();
      var sortingClasses__soy598 = '';
      sortingClasses__soy598 += 'nav-link nav-link-monospaced';
      sortingClasses__soy598 += sortingOrder == 'desc' ? ' order-arrow-down-active' : ' order-arrow-up-active';
      if (sortingURL && !disabled) {
        var msg_4424__soy4425 = '';
        /** @desc  */
        var MSG_EXTERNAL_6181231678531594473 = goog.getMsg('reverse-sort-direction');
        msg_4424__soy4425 += MSG_EXTERNAL_6181231678531594473;
        $templateAlias3({ elementClasses: sortingClasses__soy598, events: { click: _handleSortingButtonClicked }, href: sortingURL, icon: 'order-arrow', ref: 'sortingButton', spritemap: spritemap, title: '' + msg_4424__soy4425 }, null, opt_ijData);
      } else {
        var msg_4427__soy4428 = '';
        /** @desc  */
        var MSG_EXTERNAL_6181231678531594473$$1 = goog.getMsg('reverse-sort-direction');
        msg_4427__soy4428 += MSG_EXTERNAL_6181231678531594473$$1;
        $templateAlias4({ disabled: disabled, elementClasses: sortingClasses__soy598, events: { click: _handleSortingButtonClicked }, icon: 'order-arrow', ref: 'sortingButton', spritemap: spritemap, style: 'unstyled', title: '' + msg_4427__soy4428 }, null, opt_ijData);
      }
      incrementalDom.elementClose('li');
    }
    incrementalDom.elementClose('ul');
    if (showSearch) {
      var searchClasses__soy648 = '';
      searchClasses__soy648 += 'navbar-form navbar-form-autofit';
      searchClasses__soy648 += !onlySearch__soy537 ? ' navbar-overlay navbar-overlay-sm-down' + (_showSearchMobile ? ' show' : '') : '';
      incrementalDom.elementOpenStart('div');
      incrementalDom.attr('class', searchClasses__soy648);
      incrementalDom.elementOpenEnd();
      var param670 = function param670() {
        $searchInput({ _handleCloseMobileSearchClick: _handleCloseMobileSearchClick, _handleSearchSearchClick: _handleSearchSearchClick, contentRenderer: contentRenderer, disabled: disabled && !searchValue, onlySearch: onlySearch__soy537, searchInputName: searchInputName, searchInputPlaceholder: searchInputPlaceholder, searchValue: searchValue, showAdvancedSearch: showAdvancedSearch, spritemap: spritemap }, null, opt_ijData);
      };
      soy.$$getDelegateFn(soy.$$getDelTemplateId('ClayManagementToolbar.SearchForm.idom'), contentRenderer, false)({ _handleCloseMobileSearchClick: _handleCloseMobileSearchClick, _handleSearchSearchClick: _handleSearchSearchClick, customData: customData, disabled: disabled && !searchValue, onlySearch: onlySearch__soy537, searchActionURL: searchActionURL, searchFormMethod: searchFormMethod, searchFormName: searchFormName, searchInput: param670, searchInputName: searchInputName, searchValue: searchValue, spritemap: spritemap }, null, opt_ijData);
      incrementalDom.elementClose('div');
    }
    incrementalDom.elementOpenStart('ul');
    incrementalDom.attr('class', 'navbar-nav');
    incrementalDom.elementOpenEnd();
    if (showSearch && !onlySearch__soy537) {
      incrementalDom.elementOpenStart('li');
      incrementalDom.attr('class', 'nav-item navbar-breakpoint-d-none');
      incrementalDom.elementOpenEnd();
      $templateAlias4({ disabled: disabled, elementClasses: 'nav-link nav-link-monospaced', events: { click: _handleOpenMobileSearchClick }, icon: 'search', spritemap: spritemap, style: 'unstyled' }, null, opt_ijData);
      incrementalDom.elementClose('li');
    }
    if (showInfoButton) {
      var msg_4430__soy4431 = '';
      /** @desc  */
      var MSG_EXTERNAL_4279081882680795350 = goog.getMsg('info');
      msg_4430__soy4431 += MSG_EXTERNAL_4279081882680795350;
      incrementalDom.elementOpenStart('li');
      incrementalDom.attr('class', 'nav-item');
      incrementalDom.elementOpenEnd();
      $templateAlias4({ elementClasses: 'nav-link nav-link-monospaced', events: { click: _handleInfoButtonClicked }, icon: 'info-circle-open', ref: 'infoButton', spritemap: spritemap, style: 'unstyled', title: '' + msg_4430__soy4431 }, null, opt_ijData);
      incrementalDom.elementClose('li');
    }
    if (viewTypes) {
      var msg_4433__soy4434 = '';
      /** @desc  */
      var MSG_EXTERNAL_7058281898030636028 = goog.getMsg('display-style');
      msg_4433__soy4434 += MSG_EXTERNAL_7058281898030636028;
      incrementalDom.elementOpenStart('li');
      incrementalDom.attr('class', 'dropdown nav-item');
      incrementalDom.elementOpenEnd();
      var param725 = '';
      var viewType734List = viewTypes;
      var viewType734ListLen = viewType734List.length;
      for (var viewType734Index = 0; viewType734Index < viewType734ListLen; viewType734Index++) {
        var viewType734Data = viewType734List[viewType734Index];
        param725 += viewType734Data.active ? viewType734Data.icon : '';
      }
      $templateAlias9({ events: { itemClicked: _handleViewTypeClicked }, icon: param725, items: viewTypes, itemsIconAlignment: 'left', ref: 'viewTypesDropdown', spritemap: spritemap, triggerClasses: 'nav-link nav-link-monospaced', triggerTitle: '' + msg_4433__soy4434 }, null, opt_ijData);
      incrementalDom.elementClose('li');
    }
    if (showCreationMenu) {
      incrementalDom.elementOpenStart('li');
      incrementalDom.attr('class', 'nav-item');
      incrementalDom.elementOpenEnd();
      $creationMenu(opt_data, null, opt_ijData);
      incrementalDom.elementClose('li');
    }
    incrementalDom.elementClose('ul');
    incrementalDom.elementClose('div');
    incrementalDom.elementClose('nav');
  }
  exports.default = $default;
  /**
   * @typedef {{
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleCloseMobileSearchClick: (*|null|undefined),
   *  _handleCreationButtonClicked: (*|null|undefined),
   *  _handleCreationMenuItemClicked: (*|null|undefined),
   *  _handleCreationMenuMoreButtonClicked: (*|null|undefined),
   *  _handleFilterDoneButtonClick: (*|null|undefined),
   *  _handleFilterItemClicked: (*|null|undefined),
   *  _handleInfoButtonClicked: (*|null|undefined),
   *  _handleOpenMobileSearchClick: (*|null|undefined),
   *  _handleSearchSearchClick: (*|null|undefined),
   *  _handleSelectPageCheckboxChanged: (*|null|undefined),
   *  _handleSortingButtonClicked: (*|null|undefined),
   *  _handleViewTypeClicked: (*|null|undefined),
   *  _showSearchMobile: (boolean|null|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  creationMenu: (null|undefined|{caption: (!goog.soy.data.SanitizedContent|string), helpText: (!goog.soy.data.SanitizedContent|string), itemsIconAlignment: (!goog.soy.data.SanitizedContent|string), maxPrimaryItems: number, maxSecondaryItems: number, maxTotalItems: number, primaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, secondaryItems: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), items: !Array<{data: *, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean}>, viewMoreURL: (!goog.soy.data.SanitizedContent|string)}),
   *  customData: (?),
   *  disabled: (boolean|null|undefined),
   *  filterItems: (!Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), items: !Array<{active: boolean, checked: boolean, disabled: boolean, inputName: (!goog.soy.data.SanitizedContent|string), inputValue: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>, label: (!goog.soy.data.SanitizedContent|string), separator: boolean, type: (!goog.soy.data.SanitizedContent|string)}>|null|undefined),
   *  searchActionURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormMethod: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInputPlaceholder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchValue: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  selectable: (boolean|null|undefined),
   *  showAdvancedSearch: (boolean|null|undefined),
   *  showCreationMenu: (boolean|null|undefined),
   *  showFiltersDoneButton: (boolean|null|undefined),
   *  showInfoButton: (boolean|null|undefined),
   *  showSearch: (boolean|null|undefined),
   *  sortingOrder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  sortingURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  viewTypes: (!Array<{active: boolean, disabled: boolean, href: (!goog.soy.data.SanitizedContent|string), icon: (!goog.soy.data.SanitizedContent|string), label: (!goog.soy.data.SanitizedContent|string)}>|null|undefined)
   * }}
   */
  $default.Params;
  if (goog.DEBUG) {
    $default.soyTemplateName = 'ClayManagementToolbar.default';
  }

  /**
   * @param {{
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleCloseMobileSearchClick: (*|null|undefined),
   *  _handleSearchSearchClick: (*|null|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  disabled: (boolean|null|undefined),
   *  onlySearch: (boolean|null|undefined),
   *  searchInputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInputPlaceholder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchValue: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  showAdvancedSearch: (boolean|null|undefined)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function $searchInput(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    /** @type {!goog.soy.data.SanitizedContent|string} */
    var spritemap = soy.asserts.assertType(goog.isString(opt_data.spritemap) || opt_data.spritemap instanceof goog.soy.data.SanitizedContent, 'spritemap', opt_data.spritemap, '!goog.soy.data.SanitizedContent|string');
    /** @type {*|null|undefined} */
    var _handleCloseMobileSearchClick = opt_data._handleCloseMobileSearchClick;
    /** @type {*|null|undefined} */
    var _handleSearchSearchClick = opt_data._handleSearchSearchClick;
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var contentRenderer = soy.asserts.assertType(opt_data.contentRenderer == null || goog.isString(opt_data.contentRenderer) || opt_data.contentRenderer instanceof goog.soy.data.SanitizedContent, 'contentRenderer', opt_data.contentRenderer, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {boolean|null|undefined} */
    var disabled = soy.asserts.assertType(opt_data.disabled == null || goog.isBoolean(opt_data.disabled) || opt_data.disabled === 1 || opt_data.disabled === 0, 'disabled', opt_data.disabled, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var onlySearch = soy.asserts.assertType(opt_data.onlySearch == null || goog.isBoolean(opt_data.onlySearch) || opt_data.onlySearch === 1 || opt_data.onlySearch === 0, 'onlySearch', opt_data.onlySearch, 'boolean|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchInputName = soy.asserts.assertType(opt_data.searchInputName == null || goog.isString(opt_data.searchInputName) || opt_data.searchInputName instanceof goog.soy.data.SanitizedContent, 'searchInputName', opt_data.searchInputName, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchInputPlaceholder = soy.asserts.assertType(opt_data.searchInputPlaceholder == null || goog.isString(opt_data.searchInputPlaceholder) || opt_data.searchInputPlaceholder instanceof goog.soy.data.SanitizedContent, 'searchInputPlaceholder', opt_data.searchInputPlaceholder, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchValue = soy.asserts.assertType(opt_data.searchValue == null || goog.isString(opt_data.searchValue) || opt_data.searchValue instanceof goog.soy.data.SanitizedContent, 'searchValue', opt_data.searchValue, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {boolean|null|undefined} */
    var showAdvancedSearch = soy.asserts.assertType(opt_data.showAdvancedSearch == null || goog.isBoolean(opt_data.showAdvancedSearch) || opt_data.showAdvancedSearch === 1 || opt_data.showAdvancedSearch === 0, 'showAdvancedSearch', opt_data.showAdvancedSearch, 'boolean|null|undefined');
    incrementalDom.elementOpenStart('div');
    incrementalDom.attr('class', 'input-group');
    incrementalDom.elementOpenEnd();
    incrementalDom.elementOpenStart('div');
    incrementalDom.attr('class', 'input-group-item');
    incrementalDom.elementOpenEnd();
    var inputAttributes__soy775 = function inputAttributes__soy775() {
      incrementalDom.attr('aria-label', 'Search');
      if (disabled) {
        incrementalDom.attr('disabled', '');
      }
      incrementalDom.attr('class', 'form-control input-group-inset input-group-inset-after');
      incrementalDom.attr('name', searchInputName);
      if (searchInputPlaceholder) {
        incrementalDom.attr('placeholder', searchInputPlaceholder);
      } else {
        var msg_4436__soy4437 = '';
        /** @desc Search text input placeholder */
        var MSG_EXTERNAL_7283410932409447729 = goog.getMsg('search-for');
        msg_4436__soy4437 += goog.string.unescapeEntities(MSG_EXTERNAL_7283410932409447729);
        incrementalDom.attr('placeholder', msg_4436__soy4437);
      }
      incrementalDom.attr('ref', 'search');
      incrementalDom.attr('type', 'text');
      if (searchValue) {
        incrementalDom.attr('value', searchValue);
      } else {
        incrementalDom.attr('value', '');
      }
    };
    incrementalDom.elementOpenStart('input');
    inputAttributes__soy775();
    incrementalDom.elementOpenEnd();
    incrementalDom.elementClose('input');
    incrementalDom.elementOpenStart('span');
    incrementalDom.attr('class', 'input-group-inset-item input-group-inset-item-after');
    incrementalDom.elementOpenEnd();
    if (!onlySearch) {
      $templateAlias4({ elementClasses: 'navbar-breakpoint-d-none', events: { click: _handleCloseMobileSearchClick }, icon: 'times', spritemap: spritemap, style: 'unstyled' }, null, opt_ijData);
    }
    $templateAlias4({ disabled: disabled, events: { click: _handleSearchSearchClick }, icon: 'search', ref: 'searchButton', spritemap: spritemap, style: 'unstyled', type: 'submit' }, null, opt_ijData);
    if (showAdvancedSearch) {
      $templateAlias9({ contentRenderer: contentRenderer, icon: 'angle-down', items: [], spritemap: spritemap }, null, opt_ijData);
    }
    incrementalDom.elementClose('span');
    incrementalDom.elementClose('div');
    incrementalDom.elementClose('div');
  }
  exports.searchInput = $searchInput;
  /**
   * @typedef {{
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleCloseMobileSearchClick: (*|null|undefined),
   *  _handleSearchSearchClick: (*|null|undefined),
   *  contentRenderer: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  disabled: (boolean|null|undefined),
   *  onlySearch: (boolean|null|undefined),
   *  searchInputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInputPlaceholder: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchValue: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  showAdvancedSearch: (boolean|null|undefined)
   * }}
   */
  $searchInput.Params;
  if (goog.DEBUG) {
    $searchInput.soyTemplateName = 'ClayManagementToolbar.searchInput';
  }

  /**
   * @param {{
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleCloseMobileSearchClick: (*|null|undefined),
   *  _handleSearchSearchClick: (*|null|undefined),
   *  customData: (?),
   *  disabled: (boolean|null|undefined),
   *  onlySearch: (boolean|null|undefined),
   *  searchActionURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormMethod: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInput: (function()|null|undefined),
   *  searchInputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchValue: (!goog.soy.data.SanitizedContent|null|string|undefined)
   * }} opt_data
   * @param {Object<string, *>=} opt_ijData
   * @param {Object<string, *>=} opt_ijData_deprecated
   * @return {void}
   * @suppress {checkTypes}
   */
  function __deltemplate_s837_ca02a069(opt_data, opt_ijData, opt_ijData_deprecated) {
    opt_ijData = opt_ijData_deprecated || opt_ijData;
    var $$temp;
    /** @type {!goog.soy.data.SanitizedContent|string} */
    var spritemap = soy.asserts.assertType(goog.isString(opt_data.spritemap) || opt_data.spritemap instanceof goog.soy.data.SanitizedContent, 'spritemap', opt_data.spritemap, '!goog.soy.data.SanitizedContent|string');
    /** @type {*|null|undefined} */
    var _handleCloseMobileSearchClick = opt_data._handleCloseMobileSearchClick;
    /** @type {*|null|undefined} */
    var _handleSearchSearchClick = opt_data._handleSearchSearchClick;
    /** @type {?} */
    var customData = opt_data.customData;
    /** @type {boolean|null|undefined} */
    var disabled = soy.asserts.assertType(opt_data.disabled == null || goog.isBoolean(opt_data.disabled) || opt_data.disabled === 1 || opt_data.disabled === 0, 'disabled', opt_data.disabled, 'boolean|null|undefined');
    /** @type {boolean|null|undefined} */
    var onlySearch = soy.asserts.assertType(opt_data.onlySearch == null || goog.isBoolean(opt_data.onlySearch) || opt_data.onlySearch === 1 || opt_data.onlySearch === 0, 'onlySearch', opt_data.onlySearch, 'boolean|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchActionURL = soy.asserts.assertType(opt_data.searchActionURL == null || goog.isString(opt_data.searchActionURL) || opt_data.searchActionURL instanceof goog.soy.data.SanitizedContent, 'searchActionURL', opt_data.searchActionURL, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchFormMethod = soy.asserts.assertType(opt_data.searchFormMethod == null || goog.isString(opt_data.searchFormMethod) || opt_data.searchFormMethod instanceof goog.soy.data.SanitizedContent, 'searchFormMethod', opt_data.searchFormMethod, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchFormName = soy.asserts.assertType(opt_data.searchFormName == null || goog.isString(opt_data.searchFormName) || opt_data.searchFormName instanceof goog.soy.data.SanitizedContent, 'searchFormName', opt_data.searchFormName, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {function()|null|undefined} */
    var searchInput = soy.asserts.assertType(opt_data.searchInput == null || goog.isFunction(opt_data.searchInput), 'searchInput', opt_data.searchInput, 'function()|null|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchInputName = soy.asserts.assertType(opt_data.searchInputName == null || goog.isString(opt_data.searchInputName) || opt_data.searchInputName instanceof goog.soy.data.SanitizedContent, 'searchInputName', opt_data.searchInputName, '!goog.soy.data.SanitizedContent|null|string|undefined');
    /** @type {!goog.soy.data.SanitizedContent|null|string|undefined} */
    var searchValue = soy.asserts.assertType(opt_data.searchValue == null || goog.isString(opt_data.searchValue) || opt_data.searchValue instanceof goog.soy.data.SanitizedContent, 'searchValue', opt_data.searchValue, '!goog.soy.data.SanitizedContent|null|string|undefined');
    var searchFormContent__soy851 = function searchFormContent__soy851() {
      var searchFormAttributes__soy853 = function searchFormAttributes__soy853() {
        if (searchActionURL) {
          incrementalDom.attr('action', searchActionURL);
        }
        incrementalDom.attr('method', ($$temp = searchFormMethod) == null ? 'GET' : $$temp);
        if (searchFormName) {
          incrementalDom.attr('name', searchFormName);
        }
        incrementalDom.attr('role', 'search');
      };
      incrementalDom.elementOpenStart('form');
      searchFormAttributes__soy853();
      incrementalDom.elementOpenEnd();
      soyIdom.print(searchInput);
      incrementalDom.elementClose('form');
    };
    if (onlySearch) {
      searchFormContent__soy851();
    } else {
      incrementalDom.elementOpenStart('div');
      incrementalDom.attr('class', 'container-fluid container-fluid-max-xl');
      incrementalDom.elementOpenEnd();
      searchFormContent__soy851();
      incrementalDom.elementClose('div');
    }
  }
  exports.__deltemplate_s837_ca02a069 = __deltemplate_s837_ca02a069;
  /**
   * @typedef {{
   *  spritemap: (!goog.soy.data.SanitizedContent|string),
   *  _handleCloseMobileSearchClick: (*|null|undefined),
   *  _handleSearchSearchClick: (*|null|undefined),
   *  customData: (?),
   *  disabled: (boolean|null|undefined),
   *  onlySearch: (boolean|null|undefined),
   *  searchActionURL: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormMethod: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchFormName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchInput: (function()|null|undefined),
   *  searchInputName: (!goog.soy.data.SanitizedContent|null|string|undefined),
   *  searchValue: (!goog.soy.data.SanitizedContent|null|string|undefined)
   * }}
   */
  __deltemplate_s837_ca02a069.Params;
  if (goog.DEBUG) {
    __deltemplate_s837_ca02a069.soyTemplateName = 'ClayManagementToolbar.__deltemplate_s837_ca02a069';
  }
  soy.$$registerDelegateFn(soy.$$getDelTemplateId('ClayManagementToolbar.SearchForm.idom'), '', 0, __deltemplate_s837_ca02a069);

  exports.render.params = ["spritemap", "_handleActionItemClicked", "_handleClearResultsClick", "_handleClearSelectionButtonClick", "_handleCloseMobileSearchClick", "_handleCreationButtonClicked", "_handleCreationMenuItemClicked", "_handleCreationMenuMoreButtonClicked", "_handleFilterDoneButtonClick", "_handleFilterItemClicked", "_handleFilterLabelCloseClicked", "_handleInfoButtonClicked", "_handleOpenMobileSearchClick", "_handleQuickActionClicked", "_handleSearchSearchClick", "_handleSelectAllButtonClick", "_handleSelectPageCheckboxChanged", "_handleSortingButtonClicked", "_handleViewTypeClicked", "_showSearchMobile", "actionItems", "active", "checkboxStatus", "clearResultsURL", "clearSelectionURL", "contentRenderer", "creationMenu", "customData", "disabled", "elementClasses", "filterItems", "filterLabels", "id", "searchActionURL", "searchFormMethod", "searchFormName", "searchInputName", "searchInputPlaceholder", "searchValue", "selectable", "selectAllURL", "selectedItems", "showAdvancedSearch", "showCreationMenu", "showFiltersDoneButton", "showInfoButton", "showResultsBar", "showSearch", "showSelectAllButton", "sortingOrder", "sortingURL", "supportsBulkActions", "totalItems", "viewTypes"];
  exports.render.types = { "spritemap": "string", "_handleActionItemClicked": "any", "_handleClearResultsClick": "any", "_handleClearSelectionButtonClick": "any", "_handleCloseMobileSearchClick": "any", "_handleCreationButtonClicked": "any", "_handleCreationMenuItemClicked": "any", "_handleCreationMenuMoreButtonClicked": "any", "_handleFilterDoneButtonClick": "any", "_handleFilterItemClicked": "any", "_handleFilterLabelCloseClicked": "any", "_handleInfoButtonClicked": "any", "_handleOpenMobileSearchClick": "any", "_handleQuickActionClicked": "any", "_handleSearchSearchClick": "any", "_handleSelectAllButtonClick": "any", "_handleSelectPageCheckboxChanged": "any", "_handleSortingButtonClicked": "any", "_handleViewTypeClicked": "any", "_showSearchMobile": "bool", "actionItems": "list<[\n\t\tdisabled: bool,\n\t\thref: string,\n\t\ticon: string,\n\t\tlabel: string,\n\t\tquickAction: bool,\n\t\tseparator: bool,\n\t\ttype: string\n\t]>", "active": "bool", "checkboxStatus": "string", "clearResultsURL": "string", "clearSelectionURL": "string", "contentRenderer": "string", "creationMenu": "[\n\t\tcaption: string,\n\t\thelpText: string,\n\t\titemsIconAlignment: string,\n\t\tmaxPrimaryItems: int,\n\t\tmaxSecondaryItems: int,\n\t\tmaxTotalItems: int,\n\t\tprimaryItems: list<[\n\t\t\tdata: any,\n\t\t\tdisabled: bool,\n\t\t\thref: string,\n\t\t\ticon: string,\n\t\t\tlabel: string,\n\t\t\tseparator: bool\n\t\t]>,\n\t\tsecondaryItems: list<[\n\t\t\tdata: any,\n\t\t\tdisabled: bool,\n\t\t\thref: string,\n\t\t\ticon: string,\n\t\t\titems: list<[\n\t\t\t\tdata: any,\n\t\t\t\tdisabled: bool,\n\t\t\t\thref: string,\n\t\t\t\ticon: string,\n\t\t\t\tlabel: string,\n\t\t\t\tseparator: bool\n\t\t\t]>,\n\t\t\tlabel: string,\n\t\t\tseparator: bool\n\t\t]>,\n\t\tviewMoreURL: string\n\t]", "customData": "?", "disabled": "bool", "elementClasses": "string", "filterItems": "list<[\n\t\tactive: bool,\n\t\tchecked: bool,\n\t\tdisabled: bool,\n\t\tinputName: string,\n\t\tinputValue: string,\n\t\titems: list<[\n\t\t\tactive: bool,\n\t\t\tchecked: bool,\n\t\t\tdisabled: bool,\n\t\t\tinputName: string,\n\t\t\tinputValue: string,\n\t\t\tlabel: string,\n\t\t\tseparator: bool,\n\t\t\ttype: string\n\t\t]>,\n\t\tlabel: string,\n\t\tseparator: bool,\n\t\ttype: string\n\t]>", "filterLabels": "list<[\n\t\tcloseable: bool,\n\t\tlabel: html|string\n\t]>", "id": "string", "searchActionURL": "string", "searchFormMethod": "string", "searchFormName": "string", "searchInputName": "string", "searchInputPlaceholder": "string", "searchValue": "string", "selectable": "bool", "selectAllURL": "string", "selectedItems": "number", "showAdvancedSearch": "bool", "showCreationMenu": "bool", "showFiltersDoneButton": "bool", "showInfoButton": "bool", "showResultsBar": "bool", "showSearch": "bool", "showSelectAllButton": "bool", "sortingOrder": "string", "sortingURL": "string", "supportsBulkActions": "bool", "totalItems": "number", "viewTypes": "list<[\n\t\tactive: bool,\n\t\tdisabled: bool,\n\t\thref: string,\n\t\ticon: string,\n\t\tlabel: string\n\t]>" };
  exports.active.params = ["actionItems", "spritemap", "_handleActionItemClicked", "_handleClearSelectionButtonClick", "_handleInfoButtonClicked", "_handleQuickActionClicked", "_handleSelectAllButtonClick", "_handleSelectPageCheckboxChanged", "checkboxStatus", "clearSelectionURL", "selectAllURL", "selectedItems", "showInfoButton", "showSelectAllButton", "supportsBulkActions", "totalItems"];
  exports.active.types = { "actionItems": "list<?>", "spritemap": "string", "_handleActionItemClicked": "any", "_handleClearSelectionButtonClick": "any", "_handleInfoButtonClicked": "any", "_handleQuickActionClicked": "any", "_handleSelectAllButtonClick": "any", "_handleSelectPageCheckboxChanged": "any", "checkboxStatus": "string", "clearSelectionURL": "string", "selectAllURL": "string", "selectedItems": "number", "showInfoButton": "bool", "showSelectAllButton": "bool", "supportsBulkActions": "bool", "totalItems": "number" };
  exports.creationMenu.params = ["spritemap", "_handleCreationButtonClicked", "_handleCreationMenuItemClicked", "_handleCreationMenuMoreButtonClicked", "creationMenu"];
  exports.creationMenu.types = { "spritemap": "string", "_handleCreationButtonClicked": "any", "_handleCreationMenuItemClicked": "any", "_handleCreationMenuMoreButtonClicked": "any", "creationMenu": "[\n\t\tcaption: string,\n\t\thelpText: string,\n\t\titemsIconAlignment: string,\n\t\tmaxPrimaryItems: int,\n\t\tmaxSecondaryItems: int,\n\t\tmaxTotalItems: int,\n\t\tprimaryItems: list<[\n\t\t\tdata: any,\n\t\t\tdisabled: bool,\n\t\t\thref: string,\n\t\t\ticon: string,\n\t\t\tlabel: string,\n\t\t\tseparator: bool\n\t\t]>,\n\t\tsecondaryItems: list<[\n\t\t\tdata: any,\n\t\t\tdisabled: bool,\n\t\t\thref: string,\n\t\t\ticon: string,\n\t\t\titems: list<[\n\t\t\t\tdata: any,\n\t\t\t\tdisabled: bool,\n\t\t\t\thref: string,\n\t\t\t\ticon: string,\n\t\t\t\tlabel: string,\n\t\t\t\tseparator: bool\n\t\t\t]>,\n\t\t\tlabel: string,\n\t\t\tseparator: bool\n\t\t]>,\n\t\tviewMoreURL: string\n\t]" };
  exports.default.params = ["spritemap", "_handleCloseMobileSearchClick", "_handleCreationButtonClicked", "_handleCreationMenuItemClicked", "_handleCreationMenuMoreButtonClicked", "_handleFilterDoneButtonClick", "_handleFilterItemClicked", "_handleInfoButtonClicked", "_handleOpenMobileSearchClick", "_handleSearchSearchClick", "_handleSelectPageCheckboxChanged", "_handleSortingButtonClicked", "_handleViewTypeClicked", "_showSearchMobile", "contentRenderer", "creationMenu", "customData", "disabled", "filterItems", "searchActionURL", "searchFormMethod", "searchFormName", "searchInputName", "searchInputPlaceholder", "searchValue", "selectable", "showAdvancedSearch", "showCreationMenu", "showFiltersDoneButton", "showInfoButton", "showSearch", "sortingOrder", "sortingURL", "viewTypes"];
  exports.default.types = { "spritemap": "string", "_handleCloseMobileSearchClick": "any", "_handleCreationButtonClicked": "any", "_handleCreationMenuItemClicked": "any", "_handleCreationMenuMoreButtonClicked": "any", "_handleFilterDoneButtonClick": "any", "_handleFilterItemClicked": "any", "_handleInfoButtonClicked": "any", "_handleOpenMobileSearchClick": "any", "_handleSearchSearchClick": "any", "_handleSelectPageCheckboxChanged": "any", "_handleSortingButtonClicked": "any", "_handleViewTypeClicked": "any", "_showSearchMobile": "bool", "contentRenderer": "string", "creationMenu": "[\n\t\tcaption: string,\n\t\thelpText: string,\n\t\titemsIconAlignment: string,\n\t\tmaxPrimaryItems: int,\n\t\tmaxSecondaryItems: int,\n\t\tmaxTotalItems: int,\n\t\tprimaryItems: list<[\n\t\t\tdata: any,\n\t\t\tdisabled: bool,\n\t\t\thref: string,\n\t\t\ticon: string,\n\t\t\tlabel: string,\n\t\t\tseparator: bool\n\t\t]>,\n\t\tsecondaryItems: list<[\n\t\t\tdata: any,\n\t\t\tdisabled: bool,\n\t\t\thref: string,\n\t\t\ticon: string,\n\t\t\titems: list<[\n\t\t\t\tdata: any,\n\t\t\t\tdisabled: bool,\n\t\t\t\thref: string,\n\t\t\t\ticon: string,\n\t\t\t\tlabel: string,\n\t\t\t\tseparator: bool\n\t\t\t]>,\n\t\t\tlabel: string,\n\t\t\tseparator: bool\n\t\t]>,\n\t\tviewMoreURL: string\n\t]", "customData": "?", "disabled": "bool", "filterItems": "list<[\n\t\tactive: bool,\n\t\tchecked: bool,\n\t\tdisabled: bool,\n\t\tinputName: string,\n\t\tinputValue: string,\n\t\titems: list<[\n\t\t\tactive: bool,\n\t\t\tchecked: bool,\n\t\t\tdisabled: bool,\n\t\t\tinputName: string,\n\t\t\tinputValue: string,\n\t\t\tlabel: string,\n\t\t\tseparator: bool,\n\t\t\ttype: string\n\t\t]>,\n\t\tlabel: string,\n\t\tseparator: bool,\n\t\ttype: string\n\t]>", "searchActionURL": "string", "searchFormMethod": "string", "searchFormName": "string", "searchInputName": "string", "searchInputPlaceholder": "string", "searchValue": "string", "selectable": "bool", "showAdvancedSearch": "bool", "showCreationMenu": "bool", "showFiltersDoneButton": "bool", "showInfoButton": "bool", "showSearch": "bool", "sortingOrder": "string", "sortingURL": "string", "viewTypes": "list<[\n\t\tactive: bool,\n\t\tdisabled: bool,\n\t\thref: string,\n\t\ticon: string,\n\t\tlabel: string\n\t]>" };
  exports.searchInput.params = ["spritemap", "_handleCloseMobileSearchClick", "_handleSearchSearchClick", "contentRenderer", "disabled", "onlySearch", "searchInputName", "searchInputPlaceholder", "searchValue", "showAdvancedSearch"];
  exports.searchInput.types = { "spritemap": "string", "_handleCloseMobileSearchClick": "any", "_handleSearchSearchClick": "any", "contentRenderer": "string", "disabled": "bool", "onlySearch": "bool", "searchInputName": "string", "searchInputPlaceholder": "string", "searchValue": "string", "showAdvancedSearch": "bool" };
  exports.templates = templates = exports;
  return exports;
});

var ClayManagementToolbar = function (_Component) {
  _inherits(ClayManagementToolbar, _Component);

  function ClayManagementToolbar() {
    _classCallCheck(this, ClayManagementToolbar);

    return _possibleConstructorReturn(this, (ClayManagementToolbar.__proto__ || Object.getPrototypeOf(ClayManagementToolbar)).apply(this, arguments));
  }

  return ClayManagementToolbar;
}(_metalComponent2.default);

_metalSoy2.default.register(ClayManagementToolbar, templates);
exports.ClayManagementToolbar = ClayManagementToolbar;
exports.templates = templates;
exports.default = templates;
/* jshint ignore:end */
