// @clayui/upper-toolbar/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _form = require("@clayui/form");

var _layout = _interopRequireDefault(require("@clayui/layout"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _excluded = ["className"],
    _excluded2 = ["children", "className", "expand"],
    _excluded3 = ["children", "className"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Input = function Input(_ref) {
  var className = _ref.className,
      otherProps = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/_react.default.createElement(Item, {
    expand: true
  }, /*#__PURE__*/_react.default.createElement(_form.ClayInput.Group, null, /*#__PURE__*/_react.default.createElement(_form.ClayInput.GroupItem, null, /*#__PURE__*/_react.default.createElement(_form.ClayInput, _extends({
    className: (0, _classnames.default)(className, 'form-control-inline'),
    type: "text"
  }, otherProps)))));
};

Input.displayName = 'ClayUpperToolbarInput';

var Item = function Item(_ref2) {
  var children = _ref2.children,
      className = _ref2.className,
      expand = _ref2.expand,
      otherProps = _objectWithoutProperties(_ref2, _excluded2);

  return /*#__PURE__*/_react.default.createElement("li", _extends({
    className: (0, _classnames.default)(className, 'tbar-item', {
      'tbar-item-expand': expand
    })
  }, otherProps), /*#__PURE__*/_react.default.createElement("div", {
    className: "tbar-section"
  }, children));
};

Item.displayName = 'ClayUpperToolbarItem';

var ClayUpperToolbar = function ClayUpperToolbar(_ref3) {
  var children = _ref3.children,
      className = _ref3.className,
      otherProps = _objectWithoutProperties(_ref3, _excluded3);

  return /*#__PURE__*/_react.default.createElement("nav", _extends({
    className: (0, _classnames.default)(className, 'component-tbar', 'subnav-tbar-light', 'tbar')
  }, otherProps), /*#__PURE__*/_react.default.createElement(_layout.default.ContainerFluid, null, /*#__PURE__*/_react.default.createElement("ul", {
    className: "tbar-nav"
  }, children)));
};

ClayUpperToolbar.displayName = 'ClayUpperToolbar';

var _default = Object.assign(ClayUpperToolbar, {
  Input: Input,
  Item: Item
});

exports.default = _default;
