// @liferay/js-toolkit-core/lib/run.js
"use strict";
/**
 * SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runPkgJsonScript = exports.runNodeModulesBin = void 0;
const child_process_1 = __importDefault(require("child_process"));
const cross_spawn_1 = __importDefault(require("cross-spawn"));
function runNodeModulesBin(script, ...args) {
    const proc = cross_spawn_1.default.sync(require.resolve(`.bin/${script}`), args, {
        stdio: 'inherit',
    });
    if (proc.error) {
        throw proc.error;
    }
    return {
        signal: proc.signal,
        status: proc.status,
    };
}
exports.runNodeModulesBin = runNodeModulesBin;
function runPkgJsonScript(project, script, ...args) {
    const pkgManager = project.pkgManager || 'npm';
    if (pkgManager !== 'yarn') {
        args = ['--'].concat(args);
    }
    const proc = child_process_1.default.spawnSync(pkgManager, ['run', script, ...args], {
        shell: true,
        stdio: 'inherit',
    });
    if (proc.error) {
        throw proc.error;
    }
    return {
        signal: proc.signal,
        status: proc.status,
    };
}
exports.runPkgJsonScript = runPkgJsonScript;
