// @liferay/js-toolkit-core/lib/file/handler/Manifest.js
"use strict";
/**
 * SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = __importDefault(require("fs-extra"));
const path_1 = __importDefault(require("path"));
/**
 * A class to hold information about processed modules and optionally dump/read
 * it to/from disk.
 */
class Manifest {
    constructor() {
        this._json = {
            packages: {},
        };
    }
    /**
     * Add a processed package entry
     * @param srcPkg the source package descriptor
     * @param destPkg the destination package descriptor
     */
    addPackage(srcPkg, destPkg) {
        if (this._json.packages[srcPkg.id] === undefined) {
            this._json.packages[srcPkg.id] = {};
        }
        const pkg = this._json.packages[srcPkg.id];
        pkg.src = {
            dir: srcPkg.dir.asPosix,
            id: srcPkg.id,
            name: srcPkg.name,
            version: srcPkg.version,
        };
        pkg.dest = {
            dir: destPkg.dir.asPosix,
            id: destPkg.id,
            name: destPkg.name,
            version: destPkg.version,
        };
    }
    addModuleFlags(pkgId, moduleName, flags) {
        if (this._json.packages[pkgId] === undefined) {
            this._json.packages[pkgId] = {};
        }
        const pkg = this._json.packages[pkgId];
        if (pkg.modules === undefined) {
            pkg.modules = {};
        }
        if (pkg.modules[moduleName] === undefined) {
            pkg.modules[moduleName] = {};
        }
        if (pkg.modules[moduleName].flags === undefined) {
            pkg.modules[moduleName].flags = {};
        }
        Object.assign(pkg.modules[moduleName].flags, flags);
    }
    /**
     * Save current manifest to a file
     * @param filePath path to file
     */
    save(filePath) {
        fs_extra_1.default.ensureDirSync(path_1.default.dirname(filePath));
        fs_extra_1.default.writeFileSync(filePath, this.toJSON());
    }
    /**
     * Return the JSON serialization of this manifest
     */
    toJSON() {
        return JSON.stringify(this._json, sortObjectKeysReplacer, 2);
    }
}
exports.default = Manifest;
/**
 * Replacer function for sorting object keys when stringifying
 */
function sortObjectKeysReplacer(key, value) {
    if (value instanceof Object && !Array.isArray(value)) {
        return Object.keys(value)
            .sort()
            .reduce((sorted, key) => {
            sorted[key] = value[key];
            return sorted;
        }, {});
    }
    else {
        return value;
    }
}
