// @clayui/breadcrumb/lib/Ellipsis.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _button = _interopRequireDefault(require("@clayui/button"));

var _dropDown = _interopRequireDefault(require("@clayui/drop-down"));

var _icon = _interopRequireDefault(require("@clayui/icon"));

var _react = _interopRequireDefault(require("react"));

var _excluded = ["items", "spritemap", "title"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Ellipsis = function Ellipsis(_ref) {
  var items = _ref.items,
      spritemap = _ref.spritemap,
      title = _ref.title,
      otherProps = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/_react.default.createElement(_dropDown.default, _extends({
    className: "breadcrumb-item",
    containerElement: "li",
    trigger: /*#__PURE__*/_react.default.createElement(_button.default, {
      "aria-label": otherProps['aria-label'],
      className: "breadcrumb-link",
      displayType: "unstyled",
      title: title
    }, /*#__PURE__*/_react.default.createElement(_icon.default, {
      spritemap: spritemap,
      symbol: "ellipsis-h"
    }), /*#__PURE__*/_react.default.createElement(_icon.default, {
      spritemap: spritemap,
      symbol: "caret-bottom"
    }))
  }, otherProps, {
    "aria-label": undefined
  }), /*#__PURE__*/_react.default.createElement(_dropDown.default.ItemList, null, items.map(function (_ref2, i) {
    var href = _ref2.href,
        label = _ref2.label,
        onClick = _ref2.onClick;
    return /*#__PURE__*/_react.default.createElement(_dropDown.default.Item, {
      href: href,
      key: "breadcrumbEllipsisItem".concat(i),
      onClick: onClick,
      title: label
    }, label);
  })));
};

var _default = Ellipsis;
exports.default = _default;
