// @clayui/breadcrumb/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _shared = require("@clayui/shared");

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _warning = _interopRequireDefault(require("warning"));

var _Ellipsis = _interopRequireDefault(require("./Ellipsis"));

var _Item = _interopRequireDefault(require("./Item"));

var _excluded = ["className", "ellipsisBuffer", "ellipsisProps", "items", "spritemap"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var findActiveItems = function findActiveItems(items) {
  return items.filter(function (item) {
    return item.active;
  });
};

var ClayBreadcrumb = function ClayBreadcrumb(_ref) {
  var className = _ref.className,
      _ref$ellipsisBuffer = _ref.ellipsisBuffer,
      ellipsisBuffer = _ref$ellipsisBuffer === void 0 ? 1 : _ref$ellipsisBuffer,
      _ref$ellipsisProps = _ref.ellipsisProps,
      ellipsisProps = _ref$ellipsisProps === void 0 ? {} : _ref$ellipsisProps,
      items = _ref.items,
      spritemap = _ref.spritemap,
      otherProps = _objectWithoutProperties(_ref, _excluded);

  "production" !== "production" ? (0, _warning.default)(findActiveItems(items).length === 1, 'ClayBreadcrumb expects at least one `active` item on `items`.') : void 0;
  var activeItems = findActiveItems(items);
  var breadCrumbItems = ellipsisBuffer ? (0, _shared.getEllipsisItems)({
    EllipsisComponent: _Ellipsis.default,
    ellipsisProps: ellipsisProps,
    items: items,
    spritemap: spritemap
  }, ellipsisBuffer, items.indexOf(activeItems[0])) : items;
  return /*#__PURE__*/_react.default.createElement("ol", _extends({}, otherProps, {
    className: (0, _classnames.default)('breadcrumb', className)
  }), breadCrumbItems.map(function (item, i) {
    return /*#__PURE__*/_react.default.isValidElement(item) ? /*#__PURE__*/_react.default.cloneElement(item, {
      key: "ellipsis".concat(i)
    }) : /*#__PURE__*/_react.default.createElement(_Item.default, {
      active: item.active,
      href: item.href,
      key: "breadcrumbItem".concat(i),
      label: item.label,
      onClick: item.onClick
    });
  }));
};

var _default = ClayBreadcrumb;
exports.default = _default;
