// @clayui/color-picker/lib/Hue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _hooks = require("./hooks");

var _util = require("./util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * SPDX-FileCopyrightText: © 2019 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */
var useIsomorphicLayoutEffect = typeof window === 'undefined' ? _react.default.useEffect : _react.default.useLayoutEffect;
/**
 * Renders Hue component
 */

var ClayColorPickerHue = function ClayColorPickerHue(_ref) {
  var _ref$value = _ref.value,
      value = _ref$value === void 0 ? 0 : _ref$value,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === void 0 ? function () {} : _ref$onChange;

  var containerRef = _react.default.useRef(null);

  var selectorActive = _react.default.useRef(false);

  var _usePointerPosition = (0, _hooks.usePointerPosition)(containerRef),
      onPointerMove = _usePointerPosition.onPointerMove,
      setXY = _usePointerPosition.setXY,
      x = _usePointerPosition.x,
      y = _usePointerPosition.y;

  var removeListeners = function removeListeners() {
    selectorActive.current = false;
    window.removeEventListener('pointermove', onPointerMove);
    window.removeEventListener('pointerup', removeListeners);
  };

  useIsomorphicLayoutEffect(function () {
    if (containerRef.current && selectorActive.current) {
      onChange((0, _util.xToHue)(x, containerRef.current));
    }
  }, [x]);

  _react.default.useEffect(function () {
    if (containerRef.current) {
      setXY({
        x: (0, _util.hueToX)(value, containerRef.current),
        y: y
      });
    }
  }, [value]);

  _react.default.useEffect(function () {
    return removeListeners;
  }, []);

  return /*#__PURE__*/_react.default.createElement("div", {
    className: "clay-color-range clay-color-range-hue",
    onPointerDown: function onPointerDown(event) {
      event.preventDefault();
      selectorActive.current = true;
      onPointerMove(event);
      containerRef.current.querySelector('.clay-color-range-pointer').focus();
      window.addEventListener('pointermove', onPointerMove);
      window.addEventListener('pointerup', removeListeners);
    },
    ref: containerRef
  }, /*#__PURE__*/_react.default.createElement("button", {
    className: "clay-color-pointer clay-color-range-pointer",
    style: {
      background: "hsl(".concat(value, ", 100%, 50%)"),
      left: x - 7
    },
    type: "button"
  }));
};

var _default = ClayColorPickerHue;
exports.default = _default;
