// @clayui/multi-step-nav/lib/Indicator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _icon = _interopRequireDefault(require("@clayui/icon"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * SPDX-FileCopyrightText: © 2019 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */
var ClayMultiStepNavIndicator = /*#__PURE__*/_react.default.forwardRef(function (_ref, ref) {
  var complete = _ref.complete,
      innerRef = _ref.innerRef,
      label = _ref.label,
      onClick = _ref.onClick,
      spritemap = _ref.spritemap,
      subTitle = _ref.subTitle;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "multi-step-indicator",
    ref: ref
  }, subTitle && /*#__PURE__*/_react.default.createElement("div", {
    className: "multi-step-indicator-label"
  }, subTitle), /*#__PURE__*/_react.default.createElement("button", {
    className: "multi-step-icon",
    onClick: onClick,
    ref: innerRef,
    type: "button"
  }, complete && /*#__PURE__*/_react.default.createElement(_icon.default, {
    spritemap: spritemap,
    symbol: "check"
  }), !complete && label));
});

ClayMultiStepNavIndicator.displayName = 'ClayMultiStepNavIndicator';
var _default = ClayMultiStepNavIndicator;
exports.default = _default;
