// @clayui/multi-step-nav/lib/MultiStepNavWithBasicItems.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClayMultiStepNavWithBasicItems = void 0;

var _dropDown = require("@clayui/drop-down");

var _shared = require("@clayui/shared");

var _react = _interopRequireDefault(require("react"));

var _MultiStepNav = _interopRequireDefault(require("./MultiStepNav"));

var _excluded = ["spritemap"],
    _excluded2 = ["active", "activeIndex", "defaultActive", "maxStepsShown", "onActiveChange", "onIndexChange", "spritemap", "steps"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var MAX_STEPS_SHOWN = 7;

var IndicatorWithInnerRef = /*#__PURE__*/_react.default.forwardRef(function (_ref, ref) {
  var spritemap = _ref.spritemap,
      otherProps = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Indicator, _extends({}, otherProps, {
    innerRef: ref,
    label: "...",
    spritemap: spritemap
  }));
});

IndicatorWithInnerRef.displayName = 'ClayIndicatorWithInnerRef';

var ClayMultiStepNavWithBasicItems = function ClayMultiStepNavWithBasicItems(_ref2) {
  var active = _ref2.active,
      activeIndex = _ref2.activeIndex,
      defaultActive = _ref2.defaultActive,
      _ref2$maxStepsShown = _ref2.maxStepsShown,
      maxStepsShown = _ref2$maxStepsShown === void 0 ? MAX_STEPS_SHOWN : _ref2$maxStepsShown,
      onActiveChange = _ref2.onActiveChange,
      onIndexChange = _ref2.onIndexChange,
      spritemap = _ref2.spritemap,
      steps = _ref2.steps,
      otherProps = _objectWithoutProperties(_ref2, _excluded2);

  var _useInternalState = (0, _shared.useInternalState)({
    defaultName: 'defaultActive',
    defaultValue: defaultActive,
    handleName: 'onActiveChange',
    name: 'value',
    onChange: onActiveChange !== null && onActiveChange !== void 0 ? onActiveChange : onIndexChange,
    value: typeof active === 'undefined' ? activeIndex : active
  }),
      _useInternalState2 = _slicedToArray(_useInternalState, 2),
      internalActive = _useInternalState2[0],
      setActive = _useInternalState2[1];

  var dropdownItems;
  var showSteps = steps;
  var indexEnd = steps.length - 1;
  var lastStep = steps[indexEnd];

  if (steps.length > maxStepsShown) {
    var indexBeforeDropdown = maxStepsShown - 1;
    dropdownItems = steps.slice(indexBeforeDropdown, indexEnd).map(function (step, i) {
      var index = indexBeforeDropdown + i;
      return {
        active: internalActive === index,
        label: "".concat(index + 1, ". ").concat(step.title),
        onClick: function onClick() {
          return setActive(index);
        },
        symbolRight: internalActive > index ? 'check' : undefined
      };
    });
    showSteps = steps.slice(0, indexBeforeDropdown);
  }

  var activeInDropDown = internalActive > showSteps.length - 1 && internalActive < indexEnd;
  return /*#__PURE__*/_react.default.createElement(_MultiStepNav.default, otherProps, showSteps.map(function (_ref3, i) {
    var subTitle = _ref3.subTitle,
        title = _ref3.title;
    var complete = internalActive > i;
    return /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Item, {
      active: internalActive === i,
      complete: complete,
      expand: i + 1 !== steps.length,
      key: i
    }, /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Title, null, title), /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Divider, null), /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Indicator, {
      complete: complete,
      label: 1 + i,
      onClick: function onClick() {
        return setActive(i);
      },
      spritemap: spritemap,
      subTitle: subTitle
    }));
  }), dropdownItems && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Item, {
    active: activeInDropDown,
    complete: internalActive === indexEnd,
    expand: true
  }, /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Title, null, activeInDropDown ? steps[internalActive].title : steps[showSteps.length].title), /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Divider, null), /*#__PURE__*/_react.default.createElement(_dropDown.ClayDropDownWithItems, {
    items: dropdownItems,
    spritemap: spritemap,
    trigger: /*#__PURE__*/_react.default.createElement(IndicatorWithInnerRef, null)
  })), /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Item, {
    active: internalActive === indexEnd
  }, /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Title, null, lastStep.title), /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Divider, null), /*#__PURE__*/_react.default.createElement(_MultiStepNav.default.Indicator, {
    label: steps.length,
    onClick: function onClick() {
      return setActive(indexEnd);
    },
    spritemap: spritemap,
    subTitle: lastStep.subTitle
  }))));
};

exports.ClayMultiStepNavWithBasicItems = ClayMultiStepNavWithBasicItems;
