// @clayui/pagination/lib/Ellipsis.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _button = _interopRequireDefault(require("@clayui/button"));

var _dropDown = require("@clayui/drop-down");

var _shared = require("@clayui/shared");

var _react = _interopRequireDefault(require("react"));

var _excluded = ["alignmentPosition", "disabled", "disabledPages", "hrefConstructor", "items", "onPageChange"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var ClayPaginationEllipsis = function ClayPaginationEllipsis(_ref) {
  var _pages$, _pages, _pages$2, _pages2;

  var alignmentPosition = _ref.alignmentPosition,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      _ref$disabledPages = _ref.disabledPages,
      disabledPages = _ref$disabledPages === void 0 ? [] : _ref$disabledPages,
      hrefConstructor = _ref.hrefConstructor,
      _ref$items = _ref.items,
      items = _ref$items === void 0 ? [] : _ref$items,
      onPageChange = _ref.onPageChange,
      otherProps = _objectWithoutProperties(_ref, _excluded);

  var pages = disabled ? [] : items.map(function (page) {
    return {
      disabled: disabledPages.includes(page),
      href: hrefConstructor ? hrefConstructor(page) : undefined,
      label: String(page),
      onClick: function onClick() {
        return onPageChange && onPageChange(page);
      }
    };
  });
  var ariaLabel = otherProps['aria-label'] && !disabled ? (0, _shared.sub)(otherProps['aria-label'], [(_pages$ = pages[0]) === null || _pages$ === void 0 ? void 0 : _pages$.label, (_pages = pages[pages.length - 1]) === null || _pages === void 0 ? void 0 : _pages.label]) : undefined;
  var title = otherProps['title'] && !disabled ? (0, _shared.sub)(otherProps['title'], [(_pages$2 = pages[0]) === null || _pages$2 === void 0 ? void 0 : _pages$2.label, (_pages2 = pages[pages.length - 1]) === null || _pages2 === void 0 ? void 0 : _pages2.label]) : undefined;
  return /*#__PURE__*/_react.default.createElement(_dropDown.ClayDropDownWithItems, {
    alignmentPosition: alignmentPosition,
    className: "page-item",
    containerElement: "li",
    items: pages,
    trigger: /*#__PURE__*/_react.default.createElement(_button.default, _extends({}, otherProps, {
      "aria-label": ariaLabel,
      className: "page-link",
      disabled: disabled,
      displayType: "unstyled",
      title: title
    }), "...")
  });
};

var _default = ClayPaginationEllipsis;
exports.default = _default;
